// @ariakit/react-core/cjs/tag/tag-list.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _DIH46ZNHcjs = require('../__chunks/DIH46ZNH.cjs');



var _HCM6P72Qcjs = require('../__chunks/HCM6P72Q.cjs');


var _L2WE65ZQcjs = require('../__chunks/L2WE65ZQ.cjs');
require('../__chunks/2XVCJUVB.cjs');
require('../__chunks/PZ3DLQZR.cjs');
require('../__chunks/ZDWFUT2O.cjs');
require('../__chunks/T3QB4FR3.cjs');


var _JRPZ5DZLcjs = require('../__chunks/JRPZ5DZL.cjs');
require('../__chunks/YUBPT7BY.cjs');




var _QYN73JUBcjs = require('../__chunks/QYN73JUB.cjs');



var _HAHC3O6Rcjs = require('../__chunks/HAHC3O6R.cjs');
require('../__chunks/XMDZRF6Y.cjs');

// src/tag/tag-list.tsx
var _events = require('@ariakit/core/utils/events');
var _focus = require('@ariakit/core/utils/focus');
var _misc = require('@ariakit/core/utils/misc');
var _platform = require('@ariakit/core/utils/platform');
var _undo = require('@ariakit/core/utils/undo');
var _jsxruntime = require('react/jsx-runtime');
var TagName = "div";
var useTagList = _QYN73JUBcjs.createHook.call(void 0, 
  function useTagList2({ store, ...props }) {
    const context = _HCM6P72Qcjs.useTagProviderContext.call(void 0, );
    store = store || context;
    _misc.invariant.call(void 0, 
      store,
      process.env.NODE_ENV !== "production" && "TagList must receive a `store` prop or be wrapped in a TagProvider component."
    );
    const onMouseDownProp = props.onMouseDown;
    const onMouseDown = _HAHC3O6Rcjs.useEvent.call(void 0, (event) => {
      onMouseDownProp == null ? void 0 : onMouseDownProp(event);
      if (event.defaultPrevented) return;
      const target = event.target;
      const currentTarget = event.currentTarget;
      const focusableTarget = _focus.getClosestFocusable.call(void 0, target);
      const isSelfFocusable = focusableTarget === currentTarget;
      if (!isSelfFocusable && currentTarget.contains(focusableTarget)) return;
      const { inputElement } = store.getState();
      _events.queueBeforeEvent.call(void 0, event.currentTarget, "mouseup", () => {
        inputElement == null ? void 0 : inputElement.focus();
      });
    });
    const onKeyDownProp = props.onKeyDown;
    const onKeyDown = _HAHC3O6Rcjs.useEvent.call(void 0, (event) => {
      onKeyDownProp == null ? void 0 : onKeyDownProp(event);
      if (event.defaultPrevented) return;
      const pc = !_platform.isApple.call(void 0, );
      const z = event.key === "z" || event.key === "Z";
      const mod = pc ? event.ctrlKey : event.metaKey;
      const shiftZ = event.shiftKey && z || pc && event.key === "y";
      if (mod && shiftZ) {
        event.preventDefault();
        void _undo.UndoManager.redo();
      } else if (mod && z) {
        event.preventDefault();
        void _undo.UndoManager.undo();
      }
    });
    props = _HAHC3O6Rcjs.useWrapElement.call(void 0, 
      props,
      (element) => /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _HCM6P72Qcjs.TagScopedContextProvider, { value: store, children: element }),
      [store]
    );
    props = {
      ...props,
      onMouseDown,
      onKeyDown
    };
    props = _L2WE65ZQcjs.useComposite.call(void 0, { store, ...props });
    const orientation = _JRPZ5DZLcjs.useStoreState.call(void 0, 
      store,
      (state) => state.orientation === "both" ? void 0 : state.orientation
    );
    const items = _JRPZ5DZLcjs.useStoreState.call(void 0, store, (state) => state.renderedItems);
    const itemIds = items.filter((item) => !!item.value).map((item) => item.id);
    const labelId = _JRPZ5DZLcjs.useStoreState.call(void 0, store, (state) => {
      var _a;
      return (_a = state.labelElement) == null ? void 0 : _a.id;
    });
    const listboxProps = {};
    for (const key in props) {
      if (key === "role" || key.startsWith("aria-")) {
        const prop = key;
        listboxProps[prop] = props[prop];
        delete props[prop];
      }
    }
    const touchDevice = _DIH46ZNHcjs.useTouchDevice.call(void 0, );
    const children = /* @__PURE__ */ _jsxruntime.jsxs.call(void 0, _jsxruntime.Fragment, { children: [
      /* @__PURE__ */ _jsxruntime.jsx.call(void 0, 
        "div",
        {
          role: touchDevice ? "list" : "listbox",
          "aria-live": "polite",
          "aria-relevant": "all",
          "aria-atomic": true,
          "aria-labelledby": labelId,
          "aria-orientation": orientation,
          "aria-owns": itemIds.join(" "),
          ...listboxProps,
          style: { position: "fixed" }
        }
      ),
      props.children
    ] });
    props = {
      ...props,
      children
    };
    return props;
  }
);
var TagList = _QYN73JUBcjs.forwardRef.call(void 0, function TagList2(props) {
  const htmlProps = useTagList(props);
  return _QYN73JUBcjs.createElement.call(void 0, TagName, htmlProps);
});



exports.TagList = TagList; exports.useTagList = useTagList;
