// @ariakit/react-core/cjs/tag/tag.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _DIH46ZNHcjs = require('../__chunks/DIH46ZNH.cjs');




var _HCM6P72Qcjs = require('../__chunks/HCM6P72Q.cjs');


var _QNHUSFSWcjs = require('../__chunks/QNHUSFSW.cjs');
require('../__chunks/2XVCJUVB.cjs');
require('../__chunks/WLYREQI4.cjs');
require('../__chunks/PZ3DLQZR.cjs');
require('../__chunks/V75WWISJ.cjs');
require('../__chunks/ZDWFUT2O.cjs');
require('../__chunks/T3QB4FR3.cjs');
require('../__chunks/JRPZ5DZL.cjs');
require('../__chunks/YUBPT7BY.cjs');





var _QYN73JUBcjs = require('../__chunks/QYN73JUB.cjs');





var _HAHC3O6Rcjs = require('../__chunks/HAHC3O6R.cjs');
require('../__chunks/XMDZRF6Y.cjs');

// src/tag/tag.tsx
var _misc = require('@ariakit/core/utils/misc');
var _platform = require('@ariakit/core/utils/platform');
var _react = require('react');
var _jsxruntime = require('react/jsx-runtime');
var TagName = "div";
var useTag = _QYN73JUBcjs.createHook.call(void 0, function useTag2({
  store,
  value,
  getItem: getItemProp,
  removeOnKeyPress = true,
  ...props
}) {
  const context = _HCM6P72Qcjs.useTagContext.call(void 0, );
  store = store || context;
  _misc.invariant.call(void 0, 
    store,
    process.env.NODE_ENV !== "production" && "Tag must be wrapped in a TagList component."
  );
  const id = _HAHC3O6Rcjs.useId.call(void 0, props.id);
  const touchDevice = _DIH46ZNHcjs.useTouchDevice.call(void 0, );
  const getItem = _react.useCallback.call(void 0, 
    (item) => {
      const nextItem = { ...item, value };
      if (getItemProp) {
        return getItemProp(nextItem);
      }
      return nextItem;
    },
    [value, getItemProp]
  );
  const onKeyDownProp = props.onKeyDown;
  const removeOnKeyPressProp = _HAHC3O6Rcjs.useBooleanEvent.call(void 0, removeOnKeyPress);
  const onKeyDown = _HAHC3O6Rcjs.useEvent.call(void 0, (event) => {
    onKeyDownProp == null ? void 0 : onKeyDownProp(event);
    if (event.defaultPrevented) return;
    const { inputElement } = store.getState();
    const isBackspace = event.key === "Backspace";
    const isRemoveKey = isBackspace || event.key === "Delete";
    if (isRemoveKey && removeOnKeyPressProp(event)) {
      event.preventDefault();
      store.removeValue(value);
      const nextId = isBackspace ? store.previous() || store.next() : store.next();
      store.move(nextId);
    }
    const isPrintableKey = !event.ctrlKey && !event.metaKey && event.key.length === 1;
    const pc = !_platform.isApple.call(void 0, );
    const v = event.key === "v" || event.key === "V";
    const mod = pc ? event.ctrlKey : event.metaKey;
    const isPaste = mod && v;
    if (isPrintableKey || isPaste) {
      inputElement == null ? void 0 : inputElement.focus();
    }
  });
  const [removeId, setRemoveId] = _react.useState.call(void 0, );
  props = _HAHC3O6Rcjs.useWrapElement.call(void 0, 
    props,
    (element) => /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _HCM6P72Qcjs.TagValueContext.Provider, { value, children: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _HCM6P72Qcjs.TagRemoveIdContext.Provider, { value: setRemoveId, children: element }) }),
    [value]
  );
  props = {
    role: !touchDevice ? "option" : "listitem",
    children: value,
    "aria-describedby": removeId,
    ...props,
    id,
    onKeyDown
  };
  props = _QNHUSFSWcjs.useCompositeItem.call(void 0, {
    store,
    ...props,
    getItem
  });
  return props;
});
var Tag = _QYN73JUBcjs.memo.call(void 0, 
  _QYN73JUBcjs.forwardRef.call(void 0, function Tag2(props) {
    const htmlProps = useTag(props);
    return _QYN73JUBcjs.createElement.call(void 0, TagName, htmlProps);
  })
);



exports.Tag = Tag; exports.useTag = useTag;
