// @wordpress/components/build/validated-form-controls/validity-indicator.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/components/src/validated-form-controls/validity-indicator.tsx
var validity_indicator_exports = {};
__export(validity_indicator_exports, {
  ValidityIndicator: () => ValidityIndicator
});
module.exports = __toCommonJS(validity_indicator_exports);
var import_clsx = __toESM(require("clsx"));
var import_icons = require("@wordpress/icons");
var import_icon = __toESM(require("../icon/index.cjs"));
var import_spinner = __toESM(require("../spinner/index.cjs"));
var import_jsx_runtime = require("react/jsx-runtime");
function ValidityIndicator({
  type,
  message
}) {
  const ICON = {
    valid: import_icons.published,
    invalid: import_icons.error
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("p", {
    className: (0, import_clsx.default)("components-validated-control__indicator", `is-${type}`),
    children: [type === "validating" ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_spinner.default, {
      className: "components-validated-control__indicator-spinner"
    }) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_icon.default, {
      className: "components-validated-control__indicator-icon",
      icon: ICON[type],
      size: 16,
      fill: "currentColor"
    }), message]
  });
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ValidityIndicator
});
