// @ariakit/react-core/cjs/form/form-error.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _AF6NJM6Rcjs = require('../__chunks/AF6NJM6R.cjs');


var _WLYREQI4cjs = require('../__chunks/WLYREQI4.cjs');


var _JRPZ5DZLcjs = require('../__chunks/JRPZ5DZL.cjs');
require('../__chunks/YUBPT7BY.cjs');





var _QYN73JUBcjs = require('../__chunks/QYN73JUB.cjs');



var _HAHC3O6Rcjs = require('../__chunks/HAHC3O6R.cjs');
require('../__chunks/XMDZRF6Y.cjs');

// src/form/form-error.tsx
var _misc = require('@ariakit/core/utils/misc');
var _react = require('react');
var TagName = "div";
var useFormError = _QYN73JUBcjs.createHook.call(void 0, 
  function useFormError2({
    store,
    name: nameProp,
    getItem: getItemProp,
    ...props
  }) {
    const context = _AF6NJM6Rcjs.useFormContext.call(void 0, );
    store = store || context;
    _misc.invariant.call(void 0, 
      store,
      process.env.NODE_ENV !== "production" && "FormError must be wrapped in a Form component."
    );
    const id = _HAHC3O6Rcjs.useId.call(void 0, props.id);
    const ref = _react.useRef.call(void 0, null);
    const name = String(nameProp);
    const getItem = _react.useCallback.call(void 0, 
      (item) => {
        const nextItem = { ...item, id: id || item.id, name, type: "error" };
        if (getItemProp) {
          return getItemProp(nextItem);
        }
        return nextItem;
      },
      [id, name, getItemProp]
    );
    const children = _JRPZ5DZLcjs.useStoreState.call(void 0, store, () => {
      const error = store == null ? void 0 : store.getError(name);
      if (error == null) return;
      if (!(store == null ? void 0 : store.getFieldTouched(name))) return;
      return error;
    });
    props = {
      role: "alert",
      children,
      ...props,
      id,
      ref: _HAHC3O6Rcjs.useMergeRefs.call(void 0, ref, props.ref)
    };
    props = _WLYREQI4cjs.useCollectionItem.call(void 0, { store, ...props, getItem });
    return props;
  }
);
var FormError = _QYN73JUBcjs.memo.call(void 0, 
  _QYN73JUBcjs.forwardRef.call(void 0, function FormError2(props) {
    const htmlProps = useFormError(props);
    return _QYN73JUBcjs.createElement.call(void 0, TagName, htmlProps);
  })
);



exports.FormError = FormError; exports.useFormError = useFormError;
