// @ariakit/react-core/cjs/form/form-description.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _AF6NJM6Rcjs = require('../__chunks/AF6NJM6R.cjs');


var _WLYREQI4cjs = require('../__chunks/WLYREQI4.cjs');
require('../__chunks/YUBPT7BY.cjs');





var _QYN73JUBcjs = require('../__chunks/QYN73JUB.cjs');



var _HAHC3O6Rcjs = require('../__chunks/HAHC3O6R.cjs');
require('../__chunks/XMDZRF6Y.cjs');

// src/form/form-description.tsx
var _misc = require('@ariakit/core/utils/misc');
var _react = require('react');
var TagName = "div";
var useFormDescription = _QYN73JUBcjs.createHook.call(void 0, 
  function useFormDescription2({
    store,
    name: nameProp,
    getItem: getItemProp,
    ...props
  }) {
    const context = _AF6NJM6Rcjs.useFormContext.call(void 0, );
    store = store || context;
    _misc.invariant.call(void 0, 
      store,
      process.env.NODE_ENV !== "production" && "FormDescription must be wrapped in a Form component."
    );
    const id = _HAHC3O6Rcjs.useId.call(void 0, props.id);
    const ref = _react.useRef.call(void 0, null);
    const name = String(nameProp);
    const getItem = _react.useCallback.call(void 0, 
      (item) => {
        const nextItem = {
          ...item,
          id: id || item.id,
          name,
          type: "description"
        };
        if (getItemProp) {
          return getItemProp(nextItem);
        }
        return nextItem;
      },
      [id, name, getItemProp]
    );
    props = {
      ...props,
      id,
      ref: _HAHC3O6Rcjs.useMergeRefs.call(void 0, ref, props.ref)
    };
    props = _WLYREQI4cjs.useCollectionItem.call(void 0, { store, ...props, getItem });
    return props;
  }
);
var FormDescription = _QYN73JUBcjs.memo.call(void 0, 
  _QYN73JUBcjs.forwardRef.call(void 0, function FormDescription2(props) {
    const htmlProps = useFormDescription(props);
    return _QYN73JUBcjs.createElement.call(void 0, TagName, htmlProps);
  })
);



exports.FormDescription = FormDescription; exports.useFormDescription = useFormDescription;
