// @ariakit/react-core/cjs/form/form-checkbox.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _OR33H4LYcjs = require('../__chunks/OR33H4LY.cjs');


var _AF6NJM6Rcjs = require('../__chunks/AF6NJM6R.cjs');
require('../__chunks/WLYREQI4.cjs');


var _CRP5KKSOcjs = require('../__chunks/CRP5KKSO.cjs');


var _YEXWVZL3cjs = require('../__chunks/YEXWVZL3.cjs');
require('../__chunks/V75WWISJ.cjs');
require('../__chunks/ZDWFUT2O.cjs');
require('../__chunks/T3QB4FR3.cjs');
require('../__chunks/IRJEBRAW.cjs');
require('../__chunks/2MCSK5OW.cjs');
require('../__chunks/JRPZ5DZL.cjs');
require('../__chunks/YUBPT7BY.cjs');





var _QYN73JUBcjs = require('../__chunks/QYN73JUB.cjs');
require('../__chunks/HAHC3O6R.cjs');
require('../__chunks/XMDZRF6Y.cjs');

// src/form/form-checkbox.tsx
var _misc = require('@ariakit/core/utils/misc');
var TagName = "input";
var useFormCheckbox = _QYN73JUBcjs.createHook.call(void 0, 
  function useFormCheckbox2({
    store,
    name: nameProp,
    value,
    checked,
    defaultChecked,
    ...props
  }) {
    const context = _AF6NJM6Rcjs.useFormContext.call(void 0, );
    store = store || context;
    _misc.invariant.call(void 0, 
      store,
      process.env.NODE_ENV !== "production" && "FormCheckbox must be wrapped in a Form component."
    );
    const name = String(nameProp);
    const checkboxStore = _CRP5KKSOcjs.useCheckboxStore.call(void 0, {
      value: store.useValue(name),
      setValue: (value2) => store == null ? void 0 : store.setValue(name, value2)
    });
    props = _YEXWVZL3cjs.useCheckbox.call(void 0, { store: checkboxStore, value, checked, ...props });
    props = _OR33H4LYcjs.useFormControl.call(void 0, {
      store,
      name,
      "aria-labelledby": void 0,
      ...props
    });
    return props;
  }
);
var FormCheckbox = _QYN73JUBcjs.memo.call(void 0, 
  _QYN73JUBcjs.forwardRef.call(void 0, function FormCheckbox2(props) {
    const htmlProps = useFormCheckbox(props);
    return _QYN73JUBcjs.createElement.call(void 0, TagName, htmlProps);
  })
);



exports.FormCheckbox = FormCheckbox; exports.useFormCheckbox = useFormCheckbox;
