// @wordpress/components/build-module/toggle-group-control/toggle-group-control/as-button-group.mjs
// packages/components/src/toggle-group-control/toggle-group-control/as-button-group.tsx
import { useInstanceId } from "@wordpress/compose";
import { forwardRef, useMemo } from "@wordpress/element";
import { View } from "../../view/index.mjs";
import { useControlledValue } from "../../utils/index.mjs";
import ToggleGroupControlContext from "../context.mjs";
import { useComputeControlledOrUncontrolledValue } from "./utils.mjs";
import { jsx as _jsx } from "react/jsx-runtime";
function UnforwardedToggleGroupControlAsButtonGroup({
  children,
  isAdaptiveWidth,
  label,
  onChange,
  size,
  value: valueProp,
  id: idProp,
  setSelectedElement,
  ...otherProps
}, forwardedRef) {
  const generatedId = useInstanceId(ToggleGroupControlAsButtonGroup, "toggle-group-control-as-button-group");
  const baseId = idProp || generatedId;
  const {
    value,
    defaultValue
  } = useComputeControlledOrUncontrolledValue(valueProp);
  const [selectedValue, setSelectedValue] = useControlledValue({
    defaultValue,
    value,
    onChange
  });
  const groupContextValue = useMemo(() => ({
    baseId,
    value: selectedValue,
    setValue: setSelectedValue,
    isBlock: !isAdaptiveWidth,
    isDeselectable: true,
    size,
    setSelectedElement
  }), [baseId, selectedValue, setSelectedValue, isAdaptiveWidth, size, setSelectedElement]);
  return /* @__PURE__ */ _jsx(ToggleGroupControlContext.Provider, {
    value: groupContextValue,
    children: /* @__PURE__ */ _jsx(View, {
      "aria-label": label,
      ...otherProps,
      ref: forwardedRef,
      role: "group",
      children
    })
  });
}
var ToggleGroupControlAsButtonGroup = forwardRef(UnforwardedToggleGroupControlAsButtonGroup);
ToggleGroupControlAsButtonGroup.displayName = "ToggleGroupControlAsButtonGroup";
export {
  ToggleGroupControlAsButtonGroup
};
