// @ariakit/react-core/cjs/focus-trap/focus-trap-region.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _XQY3LU3Zcjs = require('../__chunks/XQY3LU3Z.cjs');
require('../__chunks/HST3DJBN.cjs');




var _QYN73JUBcjs = require('../__chunks/QYN73JUB.cjs');



var _HAHC3O6Rcjs = require('../__chunks/HAHC3O6R.cjs');
require('../__chunks/XMDZRF6Y.cjs');

// src/focus-trap/focus-trap-region.tsx
var _focus = require('@ariakit/core/utils/focus');
var _misc = require('@ariakit/core/utils/misc');
var _react = require('react');
var _jsxruntime = require('react/jsx-runtime');
var TagName = "div";
var useFocusTrapRegion = _QYN73JUBcjs.createHook.call(void 0, 
  function useFocusTrapRegion2({ enabled = false, ...props }) {
    const ref = _react.useRef.call(void 0, null);
    props = _HAHC3O6Rcjs.useWrapElement.call(void 0, 
      props,
      (element) => {
        const renderFocusTrap = () => {
          if (!enabled) return null;
          return /* @__PURE__ */ _jsxruntime.jsx.call(void 0, 
            _XQY3LU3Zcjs.FocusTrap,
            {
              onFocus: (event) => {
                const container = ref.current;
                if (!container) return;
                const tabbables = _focus.getAllTabbableIn.call(void 0, container, true);
                const first = tabbables[0];
                const last = tabbables[tabbables.length - 1];
                if (!tabbables.length) {
                  container.focus();
                  return;
                }
                if (event.relatedTarget === first) {
                  last == null ? void 0 : last.focus();
                } else {
                  first == null ? void 0 : first.focus();
                }
              }
            }
          );
        };
        return /* @__PURE__ */ _jsxruntime.jsxs.call(void 0, _jsxruntime.Fragment, { children: [
          renderFocusTrap(),
          element,
          renderFocusTrap()
        ] });
      },
      [enabled]
    );
    props = {
      ...props,
      ref: _HAHC3O6Rcjs.useMergeRefs.call(void 0, ref, props.ref)
    };
    return _misc.removeUndefinedValues.call(void 0, props);
  }
);
var FocusTrapRegion = _QYN73JUBcjs.forwardRef.call(void 0, function FocusTrapRegion2(props) {
  const htmlProps = useFocusTrapRegion(props);
  return _QYN73JUBcjs.createElement.call(void 0, TagName, htmlProps);
});



exports.FocusTrapRegion = FocusTrapRegion; exports.useFocusTrapRegion = useFocusTrapRegion;
