// @ariakit/react-core/cjs/__chunks/4NGXMMS3.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});





var _BAZBIDZKcjs = require('./BAZBIDZK.cjs');



var _4AJ2AADNcjs = require('./4AJ2AADN.cjs');



var _PZ3DLQZRcjs = require('./PZ3DLQZR.cjs');


var _QYN73JUBcjs = require('./QYN73JUB.cjs');

// src/menu/menu-context.tsx
var _react = require('react');
var menu = _QYN73JUBcjs.createStoreContext.call(void 0, 
  [_PZ3DLQZRcjs.CompositeContextProvider, _4AJ2AADNcjs.HovercardContextProvider],
  [_PZ3DLQZRcjs.CompositeScopedContextProvider, _4AJ2AADNcjs.HovercardScopedContextProvider]
);
var useMenuContext = menu.useContext;
var useMenuScopedContext = menu.useScopedContext;
var useMenuProviderContext = menu.useProviderContext;
var MenuContextProvider = menu.ContextProvider;
var MenuScopedContextProvider = menu.ScopedContextProvider;
var useMenuBarContext = _BAZBIDZKcjs.useMenubarContext;
var useMenuBarScopedContext = _BAZBIDZKcjs.useMenubarScopedContext;
var useMenuBarProviderContext = _BAZBIDZKcjs.useMenubarProviderContext;
var MenuBarContextProvider = _BAZBIDZKcjs.MenubarContextProvider;
var MenuBarScopedContextProvider = _BAZBIDZKcjs.MenubarScopedContextProvider;
var MenuItemCheckedContext = _react.createContext.call(void 0, 
  void 0
);













exports.useMenuContext = useMenuContext; exports.useMenuScopedContext = useMenuScopedContext; exports.useMenuProviderContext = useMenuProviderContext; exports.MenuContextProvider = MenuContextProvider; exports.MenuScopedContextProvider = MenuScopedContextProvider; exports.useMenuBarContext = useMenuBarContext; exports.useMenuBarScopedContext = useMenuBarScopedContext; exports.useMenuBarProviderContext = useMenuBarProviderContext; exports.MenuBarContextProvider = MenuBarContextProvider; exports.MenuBarScopedContextProvider = MenuBarScopedContextProvider; exports.MenuItemCheckedContext = MenuItemCheckedContext;
