// @ariakit/react-core/cjs/__chunks/7RTPYZZP.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _OXZTKJQEcjs = require('./OXZTKJQE.cjs');





var _QLWWO5FVcjs = require('./QLWWO5FV.cjs');


var _JRPZ5DZLcjs = require('./JRPZ5DZL.cjs');




var _QYN73JUBcjs = require('./QYN73JUB.cjs');






var _HAHC3O6Rcjs = require('./HAHC3O6R.cjs');

// src/combobox/combobox-list.tsx
var _misc = require('@ariakit/core/utils/misc');
var _react = require('react');
var _jsxruntime = require('react/jsx-runtime');
var TagName = "div";
var useComboboxList = _QYN73JUBcjs.createHook.call(void 0, 
  function useComboboxList2({ store, alwaysVisible, ...props }) {
    const scopedContext = _QLWWO5FVcjs.useComboboxScopedContext.call(void 0, true);
    const context = _QLWWO5FVcjs.useComboboxContext.call(void 0, );
    store = store || context;
    const scopedContextSameStore = !!store && store === scopedContext;
    _misc.invariant.call(void 0, 
      store,
      process.env.NODE_ENV !== "production" && "ComboboxList must receive a `store` prop or be wrapped in a ComboboxProvider component."
    );
    const ref = _react.useRef.call(void 0, null);
    const id = _HAHC3O6Rcjs.useId.call(void 0, props.id);
    const mounted = _JRPZ5DZLcjs.useStoreState.call(void 0, store, "mounted");
    const hidden = _OXZTKJQEcjs.isHidden.call(void 0, mounted, props.hidden, alwaysVisible);
    const style = hidden ? { ...props.style, display: "none" } : props.style;
    const multiSelectable = _JRPZ5DZLcjs.useStoreState.call(void 0, 
      store,
      (state) => Array.isArray(state.selectedValue)
    );
    const role = _HAHC3O6Rcjs.useAttribute.call(void 0, ref, "role", props.role);
    const isCompositeRole = role === "listbox" || role === "tree" || role === "grid";
    const ariaMultiSelectable = isCompositeRole ? multiSelectable || void 0 : void 0;
    const [hasListboxInside, setHasListboxInside] = _react.useState.call(void 0, false);
    const contentElement = _JRPZ5DZLcjs.useStoreState.call(void 0, store, "contentElement");
    _HAHC3O6Rcjs.useSafeLayoutEffect.call(void 0, () => {
      if (!mounted) return;
      const element = ref.current;
      if (!element) return;
      if (contentElement !== element) return;
      const callback = () => {
        setHasListboxInside(!!element.querySelector("[role='listbox']"));
      };
      const observer = new MutationObserver(callback);
      observer.observe(element, {
        subtree: true,
        childList: true,
        attributeFilter: ["role"]
      });
      callback();
      return () => observer.disconnect();
    }, [mounted, contentElement]);
    if (!hasListboxInside) {
      props = {
        role: "listbox",
        "aria-multiselectable": ariaMultiSelectable,
        ...props
      };
    }
    props = _HAHC3O6Rcjs.useWrapElement.call(void 0, 
      props,
      (element) => /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _QLWWO5FVcjs.ComboboxScopedContextProvider, { value: store, children: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _QLWWO5FVcjs.ComboboxListRoleContext.Provider, { value: role, children: element }) }),
      [store, role]
    );
    const setContentElement = id && (!scopedContext || !scopedContextSameStore) ? store.setContentElement : null;
    props = {
      hidden,
      ...props,
      id,
      ref: _HAHC3O6Rcjs.useMergeRefs.call(void 0, setContentElement, ref, props.ref),
      style
    };
    return _misc.removeUndefinedValues.call(void 0, props);
  }
);
var ComboboxList = _QYN73JUBcjs.forwardRef.call(void 0, function ComboboxList2(props) {
  const htmlProps = useComboboxList(props);
  return _QYN73JUBcjs.createElement.call(void 0, TagName, htmlProps);
});




exports.useComboboxList = useComboboxList; exports.ComboboxList = ComboboxList;
