// @ariakit/react-core/cjs/__chunks/BAZBIDZK.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _PZ3DLQZRcjs = require('./PZ3DLQZR.cjs');


var _QYN73JUBcjs = require('./QYN73JUB.cjs');

// src/menubar/menubar-context.tsx
var _react = require('react');
var menubar = _QYN73JUBcjs.createStoreContext.call(void 0, 
  [_PZ3DLQZRcjs.CompositeContextProvider],
  [_PZ3DLQZRcjs.CompositeScopedContextProvider]
);
var useMenubarContext = menubar.useContext;
var useMenubarScopedContext = menubar.useScopedContext;
var useMenubarProviderContext = menubar.useProviderContext;
var MenubarContextProvider = menubar.ContextProvider;
var MenubarScopedContextProvider = menubar.ScopedContextProvider;
var MenuItemCheckedContext = _react.createContext.call(void 0, 
  void 0
);








exports.useMenubarContext = useMenubarContext; exports.useMenubarScopedContext = useMenubarScopedContext; exports.useMenubarProviderContext = useMenubarProviderContext; exports.MenubarContextProvider = MenubarContextProvider; exports.MenubarScopedContextProvider = MenubarScopedContextProvider; exports.MenuItemCheckedContext = MenuItemCheckedContext;
