// @ariakit/react-core/cjs/__chunks/EZM2TZNU.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _PWFJUQMFcjs = require('./PWFJUQMF.cjs');


var _QYN73JUBcjs = require('./QYN73JUB.cjs');

// src/dialog/dialog-context.tsx
var _react = require('react');
var ctx = _QYN73JUBcjs.createStoreContext.call(void 0, 
  [_PWFJUQMFcjs.DisclosureContextProvider],
  [_PWFJUQMFcjs.DisclosureScopedContextProvider]
);
var useDialogContext = ctx.useContext;
var useDialogScopedContext = ctx.useScopedContext;
var useDialogProviderContext = ctx.useProviderContext;
var DialogContextProvider = ctx.ContextProvider;
var DialogScopedContextProvider = ctx.ScopedContextProvider;
var DialogHeadingContext = _react.createContext.call(void 0, void 0);
var DialogDescriptionContext = _react.createContext.call(void 0, void 0);









exports.useDialogContext = useDialogContext; exports.useDialogScopedContext = useDialogScopedContext; exports.useDialogProviderContext = useDialogProviderContext; exports.DialogContextProvider = DialogContextProvider; exports.DialogScopedContextProvider = DialogScopedContextProvider; exports.DialogHeadingContext = DialogHeadingContext; exports.DialogDescriptionContext = DialogDescriptionContext;
