// @ariakit/react-core/cjs/__chunks/HCM6P72Q.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _PZ3DLQZRcjs = require('./PZ3DLQZR.cjs');


var _QYN73JUBcjs = require('./QYN73JUB.cjs');

// src/tag/tag-context.tsx
var _react = require('react');
var TagValueContext = _react.createContext.call(void 0, null);
var TagRemoveIdContext = _react.createContext.call(void 0, 
  null
);
var ctx = _QYN73JUBcjs.createStoreContext.call(void 0, 
  [_PZ3DLQZRcjs.CompositeContextProvider],
  [_PZ3DLQZRcjs.CompositeScopedContextProvider]
);
var useTagContext = ctx.useContext;
var useTagScopedContext = ctx.useScopedContext;
var useTagProviderContext = ctx.useProviderContext;
var TagContextProvider = ctx.ContextProvider;
var TagScopedContextProvider = ctx.ScopedContextProvider;









exports.TagValueContext = TagValueContext; exports.TagRemoveIdContext = TagRemoveIdContext; exports.useTagContext = useTagContext; exports.useTagScopedContext = useTagScopedContext; exports.useTagProviderContext = useTagProviderContext; exports.TagContextProvider = TagContextProvider; exports.TagScopedContextProvider = TagScopedContextProvider;
