// @ariakit/react-core/cjs/__chunks/JEHJXSU4.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _V75WWISJcjs = require('./V75WWISJ.cjs');




var _QYN73JUBcjs = require('./QYN73JUB.cjs');



var _HAHC3O6Rcjs = require('./HAHC3O6R.cjs');

// src/button/button.tsx
var _dom = require('@ariakit/core/utils/dom');
var _react = require('react');
var TagName = "button";
var useButton = _QYN73JUBcjs.createHook.call(void 0, 
  function useButton2(props) {
    const ref = _react.useRef.call(void 0, null);
    const tagName = _HAHC3O6Rcjs.useTagName.call(void 0, ref, TagName);
    const [isNativeButton, setIsNativeButton] = _react.useState.call(void 0, 
      () => !!tagName && _dom.isButton.call(void 0, { tagName, type: props.type })
    );
    _react.useEffect.call(void 0, () => {
      if (!ref.current) return;
      setIsNativeButton(_dom.isButton.call(void 0, ref.current));
    }, []);
    props = {
      role: !isNativeButton && tagName !== "a" ? "button" : void 0,
      ...props,
      ref: _HAHC3O6Rcjs.useMergeRefs.call(void 0, ref, props.ref)
    };
    props = _V75WWISJcjs.useCommand.call(void 0, props);
    return props;
  }
);
var Button = _QYN73JUBcjs.forwardRef.call(void 0, function Button2(props) {
  const htmlProps = useButton(props);
  return _QYN73JUBcjs.createElement.call(void 0, TagName, htmlProps);
});




exports.useButton = useButton; exports.Button = Button;
