// @ariakit/react-core/cjs/__chunks/OJD3GZA3.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _BAZBIDZKcjs = require('./BAZBIDZK.cjs');


var _BIAKBYMPcjs = require('./BIAKBYMP.cjs');


var _L2WE65ZQcjs = require('./L2WE65ZQ.cjs');


var _JRPZ5DZLcjs = require('./JRPZ5DZL.cjs');




var _QYN73JUBcjs = require('./QYN73JUB.cjs');


var _HAHC3O6Rcjs = require('./HAHC3O6R.cjs');

// src/menubar/menubar.tsx
var _jsxruntime = require('react/jsx-runtime');
var TagName = "div";
var useMenubar = _QYN73JUBcjs.createHook.call(void 0, 
  function useMenubar2({
    store: storeProp,
    composite = true,
    orientation: orientationProp,
    virtualFocus,
    focusLoop,
    rtl,
    ...props
  }) {
    const context = _BAZBIDZKcjs.useMenubarProviderContext.call(void 0, );
    storeProp = storeProp || context;
    const store = _BIAKBYMPcjs.useMenubarStore.call(void 0, {
      store: storeProp,
      orientation: orientationProp,
      virtualFocus,
      focusLoop,
      rtl
    });
    const orientation = _JRPZ5DZLcjs.useStoreState.call(void 0, 
      store,
      (state) => !composite || state.orientation === "both" ? void 0 : state.orientation
    );
    props = _HAHC3O6Rcjs.useWrapElement.call(void 0, 
      props,
      (element) => /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _BAZBIDZKcjs.MenubarScopedContextProvider, { value: store, children: element }),
      [store]
    );
    if (composite) {
      props = {
        role: "menubar",
        "aria-orientation": orientation,
        ...props
      };
    }
    props = _L2WE65ZQcjs.useComposite.call(void 0, { store, composite, ...props });
    return props;
  }
);
var Menubar = _QYN73JUBcjs.forwardRef.call(void 0, function Menubar2(props) {
  const htmlProps = useMenubar(props);
  return _QYN73JUBcjs.createElement.call(void 0, TagName, htmlProps);
});




exports.useMenubar = useMenubar; exports.Menubar = Menubar;
