// @ariakit/react-core/cjs/__chunks/QLWWO5FV.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _D4JBVI4Fcjs = require('./D4JBVI4F.cjs');



var _PZ3DLQZRcjs = require('./PZ3DLQZR.cjs');


var _QYN73JUBcjs = require('./QYN73JUB.cjs');

// src/combobox/combobox-context.tsx
var _react = require('react');
var ComboboxListRoleContext = _react.createContext.call(void 0, 
  void 0
);
var ctx = _QYN73JUBcjs.createStoreContext.call(void 0, 
  [_D4JBVI4Fcjs.PopoverContextProvider, _PZ3DLQZRcjs.CompositeContextProvider],
  [_D4JBVI4Fcjs.PopoverScopedContextProvider, _PZ3DLQZRcjs.CompositeScopedContextProvider]
);
var useComboboxContext = ctx.useContext;
var useComboboxScopedContext = ctx.useScopedContext;
var useComboboxProviderContext = ctx.useProviderContext;
var ComboboxContextProvider = ctx.ContextProvider;
var ComboboxScopedContextProvider = ctx.ScopedContextProvider;
var ComboboxItemValueContext = _react.createContext.call(void 0, 
  void 0
);
var ComboboxItemCheckedContext = _react.createContext.call(void 0, false);










exports.ComboboxListRoleContext = ComboboxListRoleContext; exports.useComboboxContext = useComboboxContext; exports.useComboboxScopedContext = useComboboxScopedContext; exports.useComboboxProviderContext = useComboboxProviderContext; exports.ComboboxContextProvider = ComboboxContextProvider; exports.ComboboxScopedContextProvider = ComboboxScopedContextProvider; exports.ComboboxItemValueContext = ComboboxItemValueContext; exports.ComboboxItemCheckedContext = ComboboxItemCheckedContext;
