// @ariakit/react-core/cjs/__chunks/QP4OCRUY.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _D4JBVI4Fcjs = require('./D4JBVI4F.cjs');



var _PZ3DLQZRcjs = require('./PZ3DLQZR.cjs');


var _QYN73JUBcjs = require('./QYN73JUB.cjs');

// src/select/select-context.tsx
var _react = require('react');
var ctx = _QYN73JUBcjs.createStoreContext.call(void 0, 
  [_D4JBVI4Fcjs.PopoverContextProvider, _PZ3DLQZRcjs.CompositeContextProvider],
  [_D4JBVI4Fcjs.PopoverScopedContextProvider, _PZ3DLQZRcjs.CompositeScopedContextProvider]
);
var useSelectContext = ctx.useContext;
var useSelectScopedContext = ctx.useScopedContext;
var useSelectProviderContext = ctx.useProviderContext;
var SelectContextProvider = ctx.ContextProvider;
var SelectScopedContextProvider = ctx.ScopedContextProvider;
var SelectItemCheckedContext = _react.createContext.call(void 0, false);
var SelectHeadingContext = _react.createContext.call(void 0, null);









exports.useSelectContext = useSelectContext; exports.useSelectScopedContext = useSelectScopedContext; exports.useSelectProviderContext = useSelectProviderContext; exports.SelectContextProvider = SelectContextProvider; exports.SelectScopedContextProvider = SelectScopedContextProvider; exports.SelectItemCheckedContext = SelectItemCheckedContext; exports.SelectHeadingContext = SelectHeadingContext;
