// @ariakit/react-core/cjs/__chunks/TPU3Y2JW.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _QNHUSFSWcjs = require('./QNHUSFSW.cjs');


var _XGQS5HBPcjs = require('./XGQS5HBP.cjs');


var _N2EWCFNQcjs = require('./N2EWCFNQ.cjs');


var _PZ3DLQZRcjs = require('./PZ3DLQZR.cjs');


var _JRPZ5DZLcjs = require('./JRPZ5DZL.cjs');


var _QYN73JUBcjs = require('./QYN73JUB.cjs');



var _HAHC3O6Rcjs = require('./HAHC3O6R.cjs');

// src/composite/composite-item-offscreen.tsx
var _dom = require('@ariakit/core/utils/dom');
var _jsxruntime = require('react/jsx-runtime');
var TagName = "button";
function useCompositeItemOffscreen({ store, offscreenBehavior = "active", disabled, value, ...props }) {
  const context = _PZ3DLQZRcjs.useCompositeContext.call(void 0, );
  store = store || context;
  const id = _HAHC3O6Rcjs.useId.call(void 0, props.id);
  const { storeId, active, listElement, offscreenRoot } = _JRPZ5DZLcjs.useStoreStateObject.call(void 0, 
    store,
    {
      storeId: "id",
      active(state) {
        if (!state) return;
        if (!("selectedValue" in state) && "value" in state) {
          if (state.value === value) return true;
        }
        return !!id && state.activeId === id;
      },
      listElement(state) {
        if (!state) return;
        if (!("listElement" in state)) return;
        return state.listElement;
      },
      offscreenRoot(state) {
        if (props.offscreenRoot) return props.offscreenRoot;
        if (!state) return;
        if (!("contentElement" in state)) return;
        return state.contentElement || null;
      }
    }
  );
  const offscreenProps = _XGQS5HBPcjs.useCollectionItemOffscreen.call(void 0, {
    id,
    store,
    offscreenBehavior: active ? "active" : offscreenBehavior,
    ...props,
    offscreenRoot
  });
  if (!offscreenProps.active) {
    return {
      ...offscreenProps,
      children: value,
      role: _dom.getPopupItemRole.call(void 0, listElement),
      "aria-disabled": disabled || void 0,
      "data-offscreen-id": storeId
    };
  }
  return offscreenProps;
}
var CompositeItem2 = _QYN73JUBcjs.forwardRef.call(void 0, function CompositeItem3({
  offscreenBehavior,
  offscreenRoot,
  ...props
}) {
  const { active, ref, ...rest } = useCompositeItemOffscreen({
    offscreenBehavior,
    offscreenRoot,
    ...props
  });
  const allProps = { ...rest, ...props, ref: _HAHC3O6Rcjs.useMergeRefs.call(void 0, ref, props.ref) };
  if (active) {
    return /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _QNHUSFSWcjs.CompositeItem, { ...allProps });
  }
  const {
    store,
    rowId,
    preventScrollOnKeyDown,
    moveOnKeyPress,
    tabbable,
    getItem,
    ...htmlProps
  } = allProps;
  const Component = _N2EWCFNQcjs.Role[TagName];
  return /* @__PURE__ */ _jsxruntime.jsx.call(void 0, Component, { ...htmlProps });
});




exports.useCompositeItemOffscreen = useCompositeItemOffscreen; exports.CompositeItem = CompositeItem2;
