// @wordpress/components/build-module/date-time/date-time/index.mjs
// packages/components/src/date-time/date-time/index.tsx
import { forwardRef } from "@wordpress/element";
import { default as DatePicker } from "../date/index.mjs";
import { default as TimePicker } from "../time/index.mjs";
import { Wrapper } from "./styles.mjs";
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
var noop = () => {
};
function UnforwardedDateTimePicker({
  currentDate,
  is12Hour,
  dateOrder,
  isInvalidDate,
  onMonthPreviewed = noop,
  onChange,
  events,
  startOfWeek
}, ref) {
  return /* @__PURE__ */ _jsx(Wrapper, {
    ref,
    className: "components-datetime",
    spacing: 4,
    children: /* @__PURE__ */ _jsxs(_Fragment, {
      children: [/* @__PURE__ */ _jsx(TimePicker, {
        currentTime: currentDate,
        onChange,
        is12Hour,
        dateOrder
      }), /* @__PURE__ */ _jsx(DatePicker, {
        currentDate,
        onChange,
        isInvalidDate,
        events,
        onMonthPreviewed,
        startOfWeek
      })]
    })
  });
}
var DateTimePicker = forwardRef(UnforwardedDateTimePicker);
DateTimePicker.displayName = "DateTimePicker";
var date_time_default = DateTimePicker;
export {
  DatePicker,
  DateTimePicker,
  TimePicker,
  date_time_default as default
};
