// @ariakit/react-core/esm/select/select-renderer.js
"use client";
import {
  useSelectContext
} from "../__chunks/OFU5CPWY.js";
import {
  useCompositeRenderer
} from "../__chunks/6M5YGSDS.js";
import {
  getCollectionRendererItem,
  getCollectionRendererItemId
} from "../__chunks/66TYCCEM.js";
import "../__chunks/E5E7U2B6.js";
import "../__chunks/67C4K2ZC.js";
import "../__chunks/2P26HHWN.js";
import "../__chunks/57GJCRRF.js";
import {
  useStoreState
} from "../__chunks/XYGGOF6W.js";
import "../__chunks/4ODJH2Y5.js";
import {
  createElement,
  forwardRef
} from "../__chunks/ZNLX6YSC.js";
import "../__chunks/G47ZC7SD.js";
import "../__chunks/YXGXYGQX.js";

// src/select/select-renderer.tsx
import { toArray } from "@ariakit/core/utils/array";
import { useMemo } from "react";
var TagName = "div";
function getItemObject(item) {
  if (!item || typeof item !== "object") {
    return { value: `${item}` };
  }
  return item;
}
function findIndicesByValue(items, value) {
  var _a;
  const values = toArray(value);
  const indices = [];
  for (const [index, item] of items.entries()) {
    if (indices.length === values.length) break;
    const object = getItemObject(item);
    if (object.value != null && values.includes(object.value)) {
      indices.push(index);
    } else if ((_a = object.items) == null ? void 0 : _a.length) {
      const childIndices = findIndicesByValue(object.items, value);
      if (childIndices.length) {
        indices.push(index);
      }
    }
  }
  return indices;
}
function useSelectRenderer({
  store,
  orientation: orientationProp,
  persistentIndices: persistentIndicesProp,
  items: itemsProp,
  value: valueProp,
  ...props
}) {
  const context = useSelectContext();
  store = store || context;
  const items = useStoreState(store, (state) => {
    if (!state) return itemsProp;
    if (!state.mounted) return 0;
    return itemsProp != null ? itemsProp : state.items;
  });
  const value = useStoreState(store, (state) => valueProp != null ? valueProp : state == null ? void 0 : state.value);
  const valueIndices = useMemo(() => {
    if (!items) return [];
    if (value == null) return [];
    if (typeof items === "number") return [];
    if (!items.length) return [];
    return findIndicesByValue(items, value);
  }, [items, value]);
  const persistentIndices = useMemo(() => {
    if (persistentIndicesProp) {
      return [...persistentIndicesProp, ...valueIndices];
    }
    return valueIndices;
  }, [valueIndices, persistentIndicesProp]);
  return useCompositeRenderer({
    store,
    items,
    persistentIndices,
    ...props
  });
}
var SelectRenderer = forwardRef(function SelectRenderer2(props) {
  const htmlProps = useSelectRenderer(props);
  return createElement(TagName, htmlProps);
});
export {
  SelectRenderer,
  getCollectionRendererItem as getSelectRendererItem,
  getCollectionRendererItemId as getSelectRendererItemId
};
