// @ariakit/react-core/esm/composite/composite-overflow.js
"use client";
import {
  usePopover
} from "../__chunks/4IWMRS2E.js";
import "../__chunks/GKPFMPRM.js";
import "../__chunks/EKSBXE4P.js";
import "../__chunks/AOQQTIBO.js";
import "../__chunks/5M6RIVE2.js";
import "../__chunks/DPTNUR7A.js";
import "../__chunks/OQWWYG3J.js";
import "../__chunks/PONOZD7A.js";
import "../__chunks/6GXEOXGT.js";
import "../__chunks/YA3YOKXH.js";
import "../__chunks/U3R3WHDK.js";
import "../__chunks/YYYHIDZC.js";
import "../__chunks/A4D5O6EK.js";
import "../__chunks/VIQYURFN.js";
import "../__chunks/CZ4GFWYL.js";
import "../__chunks/Z5GCVBAY.js";
import "../__chunks/IGR4SXG2.js";
import "../__chunks/677M2CI3.js";
import "../__chunks/S7U6BLGA.js";
import "../__chunks/3BBVQAZ3.js";
import "../__chunks/3NDVDEB4.js";
import "../__chunks/AOUGVQZ3.js";
import "../__chunks/K2ZF5NU7.js";
import "../__chunks/63XF7ACK.js";
import "../__chunks/BDQ7RGF6.js";
import "../__chunks/3MLRVWUK.js";
import "../__chunks/YXMTOJME.js";
import "../__chunks/PIQ7VIUQ.js";
import "../__chunks/E5E7U2B6.js";
import "../__chunks/67C4K2ZC.js";
import "../__chunks/2P26HHWN.js";
import "../__chunks/C7ZLNJMM.js";
import "../__chunks/SWN3JYXT.js";
import {
  useStoreState
} from "../__chunks/XYGGOF6W.js";
import {
  createElement,
  createHook,
  forwardRef
} from "../__chunks/ZNLX6YSC.js";
import {
  useEvent
} from "../__chunks/G47ZC7SD.js";
import "../__chunks/YXGXYGQX.js";

// src/composite/composite-overflow.tsx
var TagName = "div";
var hiddenStyle = {
  opacity: 0,
  pointerEvents: "none"
};
var useCompositeOverflow = createHook(function useCompositeOverflow2({
  store,
  wrapperProps: wrapperPropsProp,
  portal = false,
  ...props
}) {
  const onFocusProp = props.onFocus;
  const onFocus = useEvent((event) => {
    onFocusProp == null ? void 0 : onFocusProp(event);
    if (event.defaultPrevented) return;
    store.show();
  });
  const mounted = useStoreState(store, "mounted");
  const getStyle = (styleProp) => mounted ? styleProp : { ...hiddenStyle, ...styleProp };
  const wrapperProps = {
    ...wrapperPropsProp,
    style: getStyle(wrapperPropsProp == null ? void 0 : wrapperPropsProp.style)
  };
  props = {
    role: "presentation",
    ...props,
    onFocus
  };
  props = usePopover({
    store,
    focusable: false,
    alwaysVisible: true,
    wrapperProps,
    portal,
    ...props
  });
  return props;
});
var CompositeOverflow = forwardRef(function CompositeOverflow2(props) {
  const htmlProps = useCompositeOverflow(props);
  return createElement(TagName, htmlProps);
});
export {
  CompositeOverflow,
  useCompositeOverflow
};
