// @ariakit/react-core/cjs/menu/menu-context.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});











var _4NGXMMS3cjs = require('../__chunks/4NGXMMS3.cjs');
require('../__chunks/BAZBIDZK.cjs');
require('../__chunks/4AJ2AADN.cjs');
require('../__chunks/D4JBVI4F.cjs');
require('../__chunks/EZM2TZNU.cjs');
require('../__chunks/PWFJUQMF.cjs');
require('../__chunks/PZ3DLQZR.cjs');
require('../__chunks/YUBPT7BY.cjs');
require('../__chunks/QYN73JUB.cjs');
require('../__chunks/HAHC3O6R.cjs');
require('../__chunks/XMDZRF6Y.cjs');












exports.MenuBarContextProvider = _4NGXMMS3cjs.MenuBarContextProvider; exports.MenuBarScopedContextProvider = _4NGXMMS3cjs.MenuBarScopedContextProvider; exports.MenuContextProvider = _4NGXMMS3cjs.MenuContextProvider; exports.MenuItemCheckedContext = _4NGXMMS3cjs.MenuItemCheckedContext; exports.MenuScopedContextProvider = _4NGXMMS3cjs.MenuScopedContextProvider; exports.useMenuBarContext = _4NGXMMS3cjs.useMenuBarContext; exports.useMenuBarProviderContext = _4NGXMMS3cjs.useMenuBarProviderContext; exports.useMenuBarScopedContext = _4NGXMMS3cjs.useMenuBarScopedContext; exports.useMenuContext = _4NGXMMS3cjs.useMenuContext; exports.useMenuProviderContext = _4NGXMMS3cjs.useMenuProviderContext; exports.useMenuScopedContext = _4NGXMMS3cjs.useMenuScopedContext;
