// @ariakit/react-core/cjs/menu/menu-item-checkbox.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _SBW7MNWEcjs = require('../__chunks/SBW7MNWE.cjs');


var _4NGXMMS3cjs = require('../__chunks/4NGXMMS3.cjs');
require('../__chunks/BAZBIDZK.cjs');
require('../__chunks/4AJ2AADN.cjs');
require('../__chunks/EO3O3B25.cjs');
require('../__chunks/QNHUSFSW.cjs');
require('../__chunks/2XVCJUVB.cjs');
require('../__chunks/WLYREQI4.cjs');
require('../__chunks/D4JBVI4F.cjs');
require('../__chunks/EZM2TZNU.cjs');
require('../__chunks/PWFJUQMF.cjs');
require('../__chunks/PZ3DLQZR.cjs');


var _CRP5KKSOcjs = require('../__chunks/CRP5KKSO.cjs');


var _YEXWVZL3cjs = require('../__chunks/YEXWVZL3.cjs');
require('../__chunks/V75WWISJ.cjs');
require('../__chunks/ZDWFUT2O.cjs');
require('../__chunks/T3QB4FR3.cjs');
require('../__chunks/IRJEBRAW.cjs');
require('../__chunks/2MCSK5OW.cjs');


var _JRPZ5DZLcjs = require('../__chunks/JRPZ5DZL.cjs');
require('../__chunks/YUBPT7BY.cjs');





var _QYN73JUBcjs = require('../__chunks/QYN73JUB.cjs');


var _HAHC3O6Rcjs = require('../__chunks/HAHC3O6R.cjs');
require('../__chunks/XMDZRF6Y.cjs');

// src/menu/menu-item-checkbox.tsx
var _misc = require('@ariakit/core/utils/misc');
var _react = require('react');
var TagName = "div";
function getPrimitiveValue(value) {
  if (Array.isArray(value)) {
    return value.toString();
  }
  return value;
}
function getValue(storeValue, value, checked) {
  if (value === void 0) {
    if (Array.isArray(storeValue)) return storeValue;
    return !!checked;
  }
  const primitiveValue = getPrimitiveValue(value);
  if (!Array.isArray(storeValue)) {
    if (checked) {
      return primitiveValue;
    }
    return storeValue === primitiveValue ? false : storeValue;
  }
  if (checked) {
    if (storeValue.includes(primitiveValue)) {
      return storeValue;
    }
    return [...storeValue, primitiveValue];
  }
  return storeValue.filter((v) => v !== primitiveValue);
}
var useMenuItemCheckbox = _QYN73JUBcjs.createHook.call(void 0, 
  function useMenuItemCheckbox2({
    store,
    name,
    value,
    checked,
    defaultChecked: defaultCheckedProp,
    hideOnClick = false,
    ...props
  }) {
    const context = _4NGXMMS3cjs.useMenuScopedContext.call(void 0, );
    store = store || context;
    _misc.invariant.call(void 0, 
      store,
      process.env.NODE_ENV !== "production" && "MenuItemCheckbox must be wrapped in a MenuList or Menu component"
    );
    const defaultChecked = _HAHC3O6Rcjs.useInitialValue.call(void 0, defaultCheckedProp);
    _react.useEffect.call(void 0, () => {
      store == null ? void 0 : store.setValue(name, (prevValue = []) => {
        if (!defaultChecked) return prevValue;
        return getValue(prevValue, value, true);
      });
    }, [store, name, value, defaultChecked]);
    _react.useEffect.call(void 0, () => {
      if (checked === void 0) return;
      store == null ? void 0 : store.setValue(name, (prevValue) => {
        return getValue(prevValue, value, checked);
      });
    }, [store, name, value, checked]);
    const checkboxStore = _CRP5KKSOcjs.useCheckboxStore.call(void 0, {
      value: _JRPZ5DZLcjs.useStoreState.call(void 0, store, (state) => state.values[name]),
      setValue(internalValue) {
        store == null ? void 0 : store.setValue(name, () => {
          if (checked === void 0) return internalValue;
          const nextValue = getValue(internalValue, value, checked);
          if (!Array.isArray(nextValue)) return nextValue;
          if (!Array.isArray(internalValue)) return nextValue;
          if (_misc.shallowEqual.call(void 0, internalValue, nextValue)) return internalValue;
          return nextValue;
        });
      }
    });
    props = {
      role: "menuitemcheckbox",
      ...props
    };
    props = _YEXWVZL3cjs.useCheckbox.call(void 0, {
      store: checkboxStore,
      name,
      value,
      checked,
      ...props
    });
    props = _SBW7MNWEcjs.useMenuItem.call(void 0, { store, hideOnClick, ...props });
    return props;
  }
);
var MenuItemCheckbox = _QYN73JUBcjs.memo.call(void 0, 
  _QYN73JUBcjs.forwardRef.call(void 0, function MenuItemCheckbox2(props) {
    const htmlProps = useMenuItemCheckbox(props);
    return _QYN73JUBcjs.createElement.call(void 0, TagName, htmlProps);
  })
);



exports.MenuItemCheckbox = MenuItemCheckbox; exports.useMenuItemCheckbox = useMenuItemCheckbox;
