// @ariakit/react-core/cjs/menu/menu-item-radio.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _SBW7MNWEcjs = require('../__chunks/SBW7MNWE.cjs');



var _4NGXMMS3cjs = require('../__chunks/4NGXMMS3.cjs');
require('../__chunks/BAZBIDZK.cjs');
require('../__chunks/4AJ2AADN.cjs');


var _KDTALU7Kcjs = require('../__chunks/KDTALU7K.cjs');
require('../__chunks/37QQG6NC.cjs');
require('../__chunks/EO3O3B25.cjs');
require('../__chunks/QNHUSFSW.cjs');
require('../__chunks/2XVCJUVB.cjs');
require('../__chunks/WLYREQI4.cjs');
require('../__chunks/D4JBVI4F.cjs');
require('../__chunks/EZM2TZNU.cjs');
require('../__chunks/PWFJUQMF.cjs');
require('../__chunks/PZ3DLQZR.cjs');
require('../__chunks/V75WWISJ.cjs');
require('../__chunks/ZDWFUT2O.cjs');
require('../__chunks/T3QB4FR3.cjs');


var _JRPZ5DZLcjs = require('../__chunks/JRPZ5DZL.cjs');
require('../__chunks/YUBPT7BY.cjs');





var _QYN73JUBcjs = require('../__chunks/QYN73JUB.cjs');



var _HAHC3O6Rcjs = require('../__chunks/HAHC3O6R.cjs');
require('../__chunks/XMDZRF6Y.cjs');

// src/menu/menu-item-radio.tsx
var _misc = require('@ariakit/core/utils/misc');
var _react = require('react');
var _jsxruntime = require('react/jsx-runtime');
var TagName = "div";
function getValue(prevValue, value, checked) {
  if (checked === void 0) return prevValue;
  if (checked) return value;
  return prevValue === value ? false : prevValue;
}
var useMenuItemRadio = _QYN73JUBcjs.createHook.call(void 0, 
  function useMenuItemRadio2({
    store,
    name,
    value,
    checked,
    onChange: onChangeProp,
    hideOnClick = false,
    ...props
  }) {
    const context = _4NGXMMS3cjs.useMenuScopedContext.call(void 0, );
    store = store || context;
    _misc.invariant.call(void 0, 
      store,
      process.env.NODE_ENV !== "production" && "MenuItemRadio must be wrapped in a MenuList or Menu component"
    );
    const defaultChecked = _HAHC3O6Rcjs.useInitialValue.call(void 0, props.defaultChecked);
    _react.useEffect.call(void 0, () => {
      store == null ? void 0 : store.setValue(name, (prevValue = false) => {
        return getValue(prevValue, value, defaultChecked);
      });
    }, [store, name, value, defaultChecked]);
    _react.useEffect.call(void 0, () => {
      if (checked === void 0) return;
      store == null ? void 0 : store.setValue(name, (prevValue) => {
        return getValue(prevValue, value, checked);
      });
    }, [store, name, value, checked]);
    const isChecked = _JRPZ5DZLcjs.useStoreState.call(void 0, 
      store,
      (state) => state.values[name] === value
    );
    props = _HAHC3O6Rcjs.useWrapElement.call(void 0, 
      props,
      (element) => /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _4NGXMMS3cjs.MenuItemCheckedContext.Provider, { value: !!isChecked, children: element }),
      [isChecked]
    );
    props = {
      role: "menuitemradio",
      ...props
    };
    props = _KDTALU7Kcjs.useRadio.call(void 0, {
      name,
      value,
      checked: isChecked,
      onChange(event) {
        onChangeProp == null ? void 0 : onChangeProp(event);
        if (event.defaultPrevented) return;
        const element = event.currentTarget;
        store == null ? void 0 : store.setValue(name, (prevValue) => {
          return getValue(prevValue, value, checked != null ? checked : element.checked);
        });
      },
      ...props
    });
    props = _SBW7MNWEcjs.useMenuItem.call(void 0, { store, hideOnClick, ...props });
    return props;
  }
);
var MenuItemRadio = _QYN73JUBcjs.memo.call(void 0, 
  _QYN73JUBcjs.forwardRef.call(void 0, function MenuItemRadio2(props) {
    const htmlProps = useMenuItemRadio(props);
    return _QYN73JUBcjs.createElement.call(void 0, TagName, htmlProps);
  })
);



exports.MenuItemRadio = MenuItemRadio; exports.useMenuItemRadio = useMenuItemRadio;
