// @ariakit/react-core/cjs/combobox/combobox-item-offscreen.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _TPU3Y2JWcjs = require('../__chunks/TPU3Y2JW.cjs');


var _6RDQYBWNcjs = require('../__chunks/6RDQYBWN.cjs');
require('../__chunks/EO3O3B25.cjs');
require('../__chunks/QNHUSFSW.cjs');
require('../__chunks/2XVCJUVB.cjs');
require('../__chunks/XGQS5HBP.cjs');


var _N2EWCFNQcjs = require('../__chunks/N2EWCFNQ.cjs');
require('../__chunks/WLYREQI4.cjs');



var _QLWWO5FVcjs = require('../__chunks/QLWWO5FV.cjs');
require('../__chunks/D4JBVI4F.cjs');
require('../__chunks/EZM2TZNU.cjs');
require('../__chunks/PWFJUQMF.cjs');
require('../__chunks/PZ3DLQZR.cjs');
require('../__chunks/V75WWISJ.cjs');
require('../__chunks/ZDWFUT2O.cjs');
require('../__chunks/T3QB4FR3.cjs');
require('../__chunks/JRPZ5DZL.cjs');
require('../__chunks/YUBPT7BY.cjs');


var _QYN73JUBcjs = require('../__chunks/QYN73JUB.cjs');


var _HAHC3O6Rcjs = require('../__chunks/HAHC3O6R.cjs');
require('../__chunks/XMDZRF6Y.cjs');

// src/combobox/combobox-item-offscreen.tsx
var _react = require('react');
var _jsxruntime = require('react/jsx-runtime');
var TagName = "div";
var itemRoleByPopupRole = {
  menu: "menuitem",
  listbox: "option",
  tree: "treeitem"
};
function getItemRole(popupRole) {
  var _a;
  const key = popupRole;
  return (_a = itemRoleByPopupRole[key]) != null ? _a : "option";
}
function useComboboxItemOffscreen({ store, value, ...props }) {
  const context = _QLWWO5FVcjs.useComboboxScopedContext.call(void 0, );
  store = store || context;
  const offscreenProps = _TPU3Y2JWcjs.useCompositeItemOffscreen.call(void 0, { store, value, ...props });
  const popupRole = _react.useContext.call(void 0, _QLWWO5FVcjs.ComboboxListRoleContext);
  if (!offscreenProps.active) {
    return {
      ...offscreenProps,
      role: getItemRole(popupRole)
    };
  }
  return offscreenProps;
}
var ComboboxItem2 = _QYN73JUBcjs.forwardRef.call(void 0, function ComboboxItem3({
  offscreenBehavior,
  offscreenRoot,
  ...props
}) {
  const { active, ref, ...rest } = useComboboxItemOffscreen({
    offscreenBehavior,
    offscreenRoot,
    ...props
  });
  const allProps = { ...rest, ...props, ref: _HAHC3O6Rcjs.useMergeRefs.call(void 0, ref, props.ref) };
  if (active) {
    return /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _6RDQYBWNcjs.ComboboxItem, { ...allProps });
  }
  const {
    store,
    value,
    hideOnClick,
    setValueOnClick,
    selectValueOnClick,
    resetValueOnSelect,
    focusOnHover,
    blurOnHoverEnd,
    moveOnKeyPress,
    getItem,
    ...htmlProps
  } = allProps;
  const Component = _N2EWCFNQcjs.Role[TagName];
  return /* @__PURE__ */ _jsxruntime.jsx.call(void 0, Component, { ...htmlProps });
});



exports.ComboboxItem = ComboboxItem2; exports.useComboboxItemOffscreen = useComboboxItemOffscreen;
