// @ariakit/react-core/cjs/combobox/combobox-popover.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _YE42HHT5cjs = require('../__chunks/YE42HHT5.cjs');


var _YZCSZZOHcjs = require('../__chunks/YZCSZZOH.cjs');
require('../__chunks/T3NAWF7A.cjs');
require('../__chunks/UI5KWBYV.cjs');
require('../__chunks/L634CRNJ.cjs');
require('../__chunks/XQY3LU3Z.cjs');
require('../__chunks/HST3DJBN.cjs');
require('../__chunks/XINXIOPH.cjs');
require('../__chunks/W32FX7DM.cjs');
require('../__chunks/EQM67Z5W.cjs');
require('../__chunks/6GRO3ZM4.cjs');
require('../__chunks/FSMQ3LQP.cjs');
require('../__chunks/NNIVHDZH.cjs');
require('../__chunks/KPQQ36MY.cjs');
require('../__chunks/JLMLKTUB.cjs');
require('../__chunks/LCWSLOYU.cjs');
require('../__chunks/V7JASX7S.cjs');
require('../__chunks/56NPFF7P.cjs');
require('../__chunks/GASMO7Y4.cjs');
require('../__chunks/DLDJYPHW.cjs');
require('../__chunks/26P4PLHI.cjs');
require('../__chunks/YVOQ2BCB.cjs');
require('../__chunks/4SKDRUBR.cjs');
require('../__chunks/AQ2LO7WV.cjs');


var _7RTPYZZPcjs = require('../__chunks/7RTPYZZP.cjs');
require('../__chunks/OXZTKJQE.cjs');
require('../__chunks/OINITWAS.cjs');
require('../__chunks/NFOCAOK7.cjs');
require('../__chunks/N2EWCFNQ.cjs');


var _QLWWO5FVcjs = require('../__chunks/QLWWO5FV.cjs');
require('../__chunks/D4JBVI4F.cjs');
require('../__chunks/EZM2TZNU.cjs');
require('../__chunks/PWFJUQMF.cjs');
require('../__chunks/PZ3DLQZR.cjs');
require('../__chunks/ZDWFUT2O.cjs');
require('../__chunks/T3QB4FR3.cjs');


var _JRPZ5DZLcjs = require('../__chunks/JRPZ5DZL.cjs');
require('../__chunks/YUBPT7BY.cjs');




var _QYN73JUBcjs = require('../__chunks/QYN73JUB.cjs');
require('../__chunks/HAHC3O6R.cjs');
require('../__chunks/XMDZRF6Y.cjs');

// src/combobox/combobox-popover.tsx
var _dom = require('@ariakit/core/utils/dom');
var _misc = require('@ariakit/core/utils/misc');
var _react = require('react');
var TagName = "div";
function isController(target, ...ids) {
  if (!target) return false;
  if ("id" in target) {
    const selector = ids.filter(Boolean).map((id) => `[aria-controls~="${id}"]`).join(", ");
    if (!selector) return false;
    return target.matches(selector);
  }
  return false;
}
var useComboboxPopover = _QYN73JUBcjs.createHook.call(void 0, 
  function useComboboxPopover2({
    store,
    modal,
    tabIndex,
    alwaysVisible,
    autoFocusOnHide = true,
    hideOnInteractOutside = true,
    ...props
  }) {
    const context = _QLWWO5FVcjs.useComboboxProviderContext.call(void 0, );
    store = store || context;
    _misc.invariant.call(void 0, 
      store,
      process.env.NODE_ENV !== "production" && "ComboboxPopover must receive a `store` prop or be wrapped in a ComboboxProvider component."
    );
    const baseElement = _JRPZ5DZLcjs.useStoreState.call(void 0, store, "baseElement");
    const hiddenByClickOutsideRef = _react.useRef.call(void 0, false);
    const treeSnapshotKey = _JRPZ5DZLcjs.useStoreState.call(void 0, 
      store.tag,
      (state) => state == null ? void 0 : state.renderedItems.length
    );
    props = _7RTPYZZPcjs.useComboboxList.call(void 0, { store, alwaysVisible, ...props });
    props = _YE42HHT5cjs.usePopover.call(void 0, {
      store,
      modal,
      alwaysVisible,
      backdrop: false,
      autoFocusOnShow: false,
      finalFocus: baseElement,
      preserveTabOrderAnchor: null,
      unstable_treeSnapshotKey: treeSnapshotKey,
      ...props,
      // When the combobox popover is modal, we make sure to include the
      // combobox input and all the combobox controls (cancel, disclosure) in
      // the list of persistent elements so they make part of the modal context,
      // allowing users to tab through them.
      getPersistentElements() {
        var _a;
        const elements = ((_a = props.getPersistentElements) == null ? void 0 : _a.call(props)) || [];
        if (!modal) return elements;
        if (!store) return elements;
        const { contentElement, baseElement: baseElement2 } = store.getState();
        if (!baseElement2) return elements;
        const doc = _dom.getDocument.call(void 0, baseElement2);
        const selectors = [];
        if (contentElement == null ? void 0 : contentElement.id) {
          selectors.push(`[aria-controls~="${contentElement.id}"]`);
        }
        if (baseElement2 == null ? void 0 : baseElement2.id) {
          selectors.push(`[aria-controls~="${baseElement2.id}"]`);
        }
        if (!selectors.length) return [...elements, baseElement2];
        const selector = selectors.join(",");
        const controlElements = doc.querySelectorAll(selector);
        return [...elements, ...controlElements];
      },
      // The combobox popover should focus on the combobox input when it hides,
      // unless the event was triggered by a click outside the popover, in which
      // case the input shouldn't be re-focused.
      autoFocusOnHide(element) {
        if (_misc.isFalsyBooleanCallback.call(void 0, autoFocusOnHide, element)) return false;
        if (hiddenByClickOutsideRef.current) {
          hiddenByClickOutsideRef.current = false;
          return false;
        }
        return true;
      },
      // Make sure we don't hide the popover when the user interacts with the
      // combobox cancel or the combobox disclosure buttons. They will have the
      // aria-controls attribute pointing to either the combobox input or the
      // combobox popover elements.
      hideOnInteractOutside(event) {
        var _a, _b;
        const state = store == null ? void 0 : store.getState();
        const contentId = (_a = state == null ? void 0 : state.contentElement) == null ? void 0 : _a.id;
        const baseId = (_b = state == null ? void 0 : state.baseElement) == null ? void 0 : _b.id;
        if (isController(event.target, contentId, baseId)) return false;
        const result = typeof hideOnInteractOutside === "function" ? hideOnInteractOutside(event) : hideOnInteractOutside;
        if (result) {
          hiddenByClickOutsideRef.current = event.type === "click";
        }
        return result;
      }
    });
    return props;
  }
);
var ComboboxPopover = _YZCSZZOHcjs.createDialogComponent.call(void 0, 
  _QYN73JUBcjs.forwardRef.call(void 0, function ComboboxPopover2(props) {
    const htmlProps = useComboboxPopover(props);
    return _QYN73JUBcjs.createElement.call(void 0, TagName, htmlProps);
  }),
  _QLWWO5FVcjs.useComboboxProviderContext
);



exports.ComboboxPopover = ComboboxPopover; exports.useComboboxPopover = useComboboxPopover;
