// @ariakit/react-core/cjs/combobox/combobox-disclosure.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _QYHLJ3IQcjs = require('../__chunks/QYHLJ3IQ.cjs');
require('../__chunks/6VLEUQCK.cjs');


var _QLWWO5FVcjs = require('../__chunks/QLWWO5FV.cjs');
require('../__chunks/D4JBVI4F.cjs');
require('../__chunks/EZM2TZNU.cjs');
require('../__chunks/PWFJUQMF.cjs');
require('../__chunks/PZ3DLQZR.cjs');
require('../__chunks/JEHJXSU4.cjs');
require('../__chunks/V75WWISJ.cjs');
require('../__chunks/ZDWFUT2O.cjs');
require('../__chunks/T3QB4FR3.cjs');


var _JRPZ5DZLcjs = require('../__chunks/JRPZ5DZL.cjs');
require('../__chunks/YUBPT7BY.cjs');




var _QYN73JUBcjs = require('../__chunks/QYN73JUB.cjs');



var _HAHC3O6Rcjs = require('../__chunks/HAHC3O6R.cjs');
require('../__chunks/XMDZRF6Y.cjs');

// src/combobox/combobox-disclosure.tsx
var _misc = require('@ariakit/core/utils/misc');
var _jsxruntime = require('react/jsx-runtime');
var TagName = "button";
var children = /* @__PURE__ */ _jsxruntime.jsx.call(void 0, 
  "svg",
  {
    "aria-hidden": "true",
    display: "block",
    fill: "none",
    stroke: "currentColor",
    strokeLinecap: "round",
    strokeLinejoin: "round",
    strokeWidth: 1.5,
    viewBox: "0 0 16 16",
    height: "1em",
    width: "1em",
    pointerEvents: "none",
    children: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, "polyline", { points: "4,6 8,10 12,6" })
  }
);
var useComboboxDisclosure = _QYN73JUBcjs.createHook.call(void 0, function useComboboxDisclosure2({ store, ...props }) {
  const context = _QLWWO5FVcjs.useComboboxProviderContext.call(void 0, );
  store = store || context;
  _misc.invariant.call(void 0, 
    store,
    process.env.NODE_ENV !== "production" && "ComboboxDisclosure must receive a `store` prop or be wrapped in a ComboboxProvider component."
  );
  const onMouseDownProp = props.onMouseDown;
  const onMouseDown = _HAHC3O6Rcjs.useEvent.call(void 0, (event) => {
    onMouseDownProp == null ? void 0 : onMouseDownProp(event);
    event.preventDefault();
    store == null ? void 0 : store.move(null);
  });
  const onClickProp = props.onClick;
  const onClick = _HAHC3O6Rcjs.useEvent.call(void 0, (event) => {
    onClickProp == null ? void 0 : onClickProp(event);
    if (event.defaultPrevented) return;
    if (!store) return;
    const { baseElement: baseElement2 } = store.getState();
    store.setDisclosureElement(baseElement2);
  });
  const baseElement = _JRPZ5DZLcjs.useStoreState.call(void 0, store, "baseElement");
  const open = _JRPZ5DZLcjs.useStoreState.call(void 0, store, "open");
  _HAHC3O6Rcjs.useSafeLayoutEffect.call(void 0, () => {
    if (!baseElement) return;
    store.setDisclosureElement(baseElement);
  }, [store, baseElement]);
  props = {
    children,
    tabIndex: -1,
    "aria-label": open ? "Hide popup" : "Show popup",
    "aria-expanded": open,
    ...props,
    onMouseDown,
    onClick
  };
  props = _QYHLJ3IQcjs.useDialogDisclosure.call(void 0, { store, ...props });
  return props;
});
var ComboboxDisclosure = _QYN73JUBcjs.forwardRef.call(void 0, function ComboboxDisclosure2(props) {
  const htmlProps = useComboboxDisclosure(props);
  return _QYN73JUBcjs.createElement.call(void 0, TagName, htmlProps);
});



exports.ComboboxDisclosure = ComboboxDisclosure; exports.useComboboxDisclosure = useComboboxDisclosure;
