// @ariakit/react-core/cjs/tab/tab.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _VHMCDKZZcjs = require('../__chunks/VHMCDKZZ.cjs');



var _QNHUSFSWcjs = require('../__chunks/QNHUSFSW.cjs');
require('../__chunks/2XVCJUVB.cjs');
require('../__chunks/WLYREQI4.cjs');
require('../__chunks/PZ3DLQZR.cjs');
require('../__chunks/V75WWISJ.cjs');
require('../__chunks/ZDWFUT2O.cjs');
require('../__chunks/T3QB4FR3.cjs');


var _JRPZ5DZLcjs = require('../__chunks/JRPZ5DZL.cjs');
require('../__chunks/YUBPT7BY.cjs');





var _QYN73JUBcjs = require('../__chunks/QYN73JUB.cjs');



var _HAHC3O6Rcjs = require('../__chunks/HAHC3O6R.cjs');
require('../__chunks/XMDZRF6Y.cjs');

// src/tab/tab.tsx
var _misc = require('@ariakit/core/utils/misc');
var _react = require('react');
var _jsxruntime = require('react/jsx-runtime');
var TagName = "button";
var useTab = _QYN73JUBcjs.createHook.call(void 0, function useTab2({
  store,
  getItem: getItemProp,
  ...props
}) {
  var _a;
  const context = _VHMCDKZZcjs.useTabScopedContext.call(void 0, );
  store = store || context;
  _misc.invariant.call(void 0, 
    store,
    process.env.NODE_ENV !== "production" && "Tab must be wrapped in a TabList component."
  );
  const defaultId = _HAHC3O6Rcjs.useId.call(void 0, );
  const id = props.id || defaultId;
  const dimmed = _misc.disabledFromProps.call(void 0, props);
  const getItem = _react.useCallback.call(void 0, 
    (item) => {
      const nextItem = { ...item, dimmed };
      if (getItemProp) {
        return getItemProp(nextItem);
      }
      return nextItem;
    },
    [dimmed, getItemProp]
  );
  const onClickProp = props.onClick;
  const onClick = _HAHC3O6Rcjs.useEvent.call(void 0, (event) => {
    onClickProp == null ? void 0 : onClickProp(event);
    if (event.defaultPrevented) return;
    store == null ? void 0 : store.setSelectedId(id);
  });
  const panelId = _JRPZ5DZLcjs.useStoreState.call(void 0, 
    store.panels,
    (state) => {
      var _a2;
      return (_a2 = state.items.find((item) => item.tabId === id)) == null ? void 0 : _a2.id;
    }
  );
  const shouldRegisterItem = defaultId ? props.shouldRegisterItem : false;
  const isActive = _JRPZ5DZLcjs.useStoreState.call(void 0, 
    store,
    (state) => !!id && state.activeId === id
  );
  const selected = _JRPZ5DZLcjs.useStoreState.call(void 0, 
    store,
    (state) => !!id && state.selectedId === id
  );
  const hasActiveItem = _JRPZ5DZLcjs.useStoreState.call(void 0, 
    store,
    (state) => !!store.item(state.activeId)
  );
  const canRegisterComposedItem = isActive || selected && !hasActiveItem;
  const accessibleWhenDisabled = selected || ((_a = props.accessibleWhenDisabled) != null ? _a : true);
  const isWithinVirtualFocusComposite = _JRPZ5DZLcjs.useStoreState.call(void 0, 
    store.combobox || store.composite,
    "virtualFocus"
  );
  if (isWithinVirtualFocusComposite) {
    props = {
      ...props,
      tabIndex: -1
    };
  }
  props = {
    role: "tab",
    "aria-selected": selected,
    "aria-controls": panelId || void 0,
    ...props,
    id,
    onClick
  };
  if (store.composite) {
    const defaultProps = {
      id,
      accessibleWhenDisabled,
      store: store.composite,
      shouldRegisterItem: canRegisterComposedItem && shouldRegisterItem,
      rowId: props.rowId,
      render: props.render
    };
    props = {
      ...props,
      render: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, 
        _QNHUSFSWcjs.CompositeItem,
        {
          ...defaultProps,
          render: store.combobox && store.composite !== store.combobox ? /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _QNHUSFSWcjs.CompositeItem, { ...defaultProps, store: store.combobox }) : defaultProps.render
        }
      )
    };
  }
  props = _QNHUSFSWcjs.useCompositeItem.call(void 0, {
    store,
    ...props,
    accessibleWhenDisabled,
    getItem,
    shouldRegisterItem
  });
  return props;
});
var Tab = _QYN73JUBcjs.memo.call(void 0, 
  _QYN73JUBcjs.forwardRef.call(void 0, function Tab2(props) {
    const htmlProps = useTab(props);
    return _QYN73JUBcjs.createElement.call(void 0, TagName, htmlProps);
  })
);



exports.Tab = Tab; exports.useTab = useTab;
