// @ariakit/react-core/cjs/toolbar/toolbar.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _VFYHXLBNcjs = require('../__chunks/VFYHXLBN.cjs');



var _QOAGU4ZScjs = require('../__chunks/QOAGU4ZS.cjs');
require('../__chunks/KTMZUVQI.cjs');


var _L2WE65ZQcjs = require('../__chunks/L2WE65ZQ.cjs');
require('../__chunks/2XVCJUVB.cjs');
require('../__chunks/NYU2UCLT.cjs');
require('../__chunks/PZ3DLQZR.cjs');
require('../__chunks/ZDWFUT2O.cjs');
require('../__chunks/T3QB4FR3.cjs');


var _JRPZ5DZLcjs = require('../__chunks/JRPZ5DZL.cjs');
require('../__chunks/YUBPT7BY.cjs');




var _QYN73JUBcjs = require('../__chunks/QYN73JUB.cjs');


var _HAHC3O6Rcjs = require('../__chunks/HAHC3O6R.cjs');
require('../__chunks/XMDZRF6Y.cjs');

// src/toolbar/toolbar.tsx
var _jsxruntime = require('react/jsx-runtime');
var TagName = "div";
var useToolbar = _QYN73JUBcjs.createHook.call(void 0, 
  function useToolbar2({
    store: storeProp,
    orientation: orientationProp,
    virtualFocus,
    focusLoop,
    rtl,
    ...props
  }) {
    const context = _QOAGU4ZScjs.useToolbarProviderContext.call(void 0, );
    storeProp = storeProp || context;
    const store = _VFYHXLBNcjs.useToolbarStore.call(void 0, {
      store: storeProp,
      orientation: orientationProp,
      virtualFocus,
      focusLoop,
      rtl
    });
    const orientation = _JRPZ5DZLcjs.useStoreState.call(void 0, 
      store,
      (state) => state.orientation === "both" ? void 0 : state.orientation
    );
    props = _HAHC3O6Rcjs.useWrapElement.call(void 0, 
      props,
      (element) => /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _QOAGU4ZScjs.ToolbarScopedContextProvider, { value: store, children: element }),
      [store]
    );
    props = {
      role: "toolbar",
      "aria-orientation": orientation,
      ...props
    };
    props = _L2WE65ZQcjs.useComposite.call(void 0, { store, ...props });
    return props;
  }
);
var Toolbar = _QYN73JUBcjs.forwardRef.call(void 0, function Toolbar2(props) {
  const htmlProps = useToolbar(props);
  return _QYN73JUBcjs.createElement.call(void 0, TagName, htmlProps);
});



exports.Toolbar = Toolbar; exports.useToolbar = useToolbar;
