// @wordpress/components/build/toggle-group-control/toggle-group-control/as-button-group.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/components/src/toggle-group-control/toggle-group-control/as-button-group.tsx
var as_button_group_exports = {};
__export(as_button_group_exports, {
  ToggleGroupControlAsButtonGroup: () => ToggleGroupControlAsButtonGroup
});
module.exports = __toCommonJS(as_button_group_exports);
var import_compose = require("@wordpress/compose");
var import_element = require("@wordpress/element");
var import_view = require("../../view/index.cjs");
var import_utils = require("../../utils/index.cjs");
var import_context = __toESM(require("../context.cjs"));
var import_utils2 = require("./utils.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
function UnforwardedToggleGroupControlAsButtonGroup({
  children,
  isAdaptiveWidth,
  label,
  onChange,
  size,
  value: valueProp,
  id: idProp,
  setSelectedElement,
  ...otherProps
}, forwardedRef) {
  const generatedId = (0, import_compose.useInstanceId)(ToggleGroupControlAsButtonGroup, "toggle-group-control-as-button-group");
  const baseId = idProp || generatedId;
  const {
    value,
    defaultValue
  } = (0, import_utils2.useComputeControlledOrUncontrolledValue)(valueProp);
  const [selectedValue, setSelectedValue] = (0, import_utils.useControlledValue)({
    defaultValue,
    value,
    onChange
  });
  const groupContextValue = (0, import_element.useMemo)(() => ({
    baseId,
    value: selectedValue,
    setValue: setSelectedValue,
    isBlock: !isAdaptiveWidth,
    isDeselectable: true,
    size,
    setSelectedElement
  }), [baseId, selectedValue, setSelectedValue, isAdaptiveWidth, size, setSelectedElement]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_context.default.Provider, {
    value: groupContextValue,
    children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_view.View, {
      "aria-label": label,
      ...otherProps,
      ref: forwardedRef,
      role: "group",
      children
    })
  });
}
var ToggleGroupControlAsButtonGroup = (0, import_element.forwardRef)(UnforwardedToggleGroupControlAsButtonGroup);
ToggleGroupControlAsButtonGroup.displayName = "ToggleGroupControlAsButtonGroup";
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ToggleGroupControlAsButtonGroup
});
