// @wordpress/components/build/card/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/components/src/card/index.ts
var card_exports = {};
__export(card_exports, {
  Card: () => import_card.default,
  CardBody: () => import_card_body.default,
  CardDivider: () => import_card_divider.default,
  CardFooter: () => import_card_footer.default,
  CardHeader: () => import_card_header.default,
  CardMedia: () => import_card_media.default,
  useCard: () => import_card.useCard,
  useCardBody: () => import_card_body.useCardBody,
  useCardDivider: () => import_card_divider.useCardDivider,
  useCardFooter: () => import_card_footer.useCardFooter,
  useCardHeader: () => import_card_header.useCardHeader,
  useCardMedia: () => import_card_media.useCardMedia
});
module.exports = __toCommonJS(card_exports);
var import_card = __toESM(require("./card/index.cjs"));
var import_card_body = __toESM(require("./card-body/index.cjs"));
var import_card_divider = __toESM(require("./card-divider/index.cjs"));
var import_card_footer = __toESM(require("./card-footer/index.cjs"));
var import_card_header = __toESM(require("./card-header/index.cjs"));
var import_card_media = __toESM(require("./card-media/index.cjs"));
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Card,
  CardBody,
  CardDivider,
  CardFooter,
  CardHeader,
  CardMedia,
  useCard,
  useCardBody,
  useCardDivider,
  useCardFooter,
  useCardHeader,
  useCardMedia
});
