// @ariakit/react-core/esm/form/form-submit.js
"use client";
import {
  useFormContext
} from "../__chunks/5QCILLRJ.js";
import {
  useButton
} from "../__chunks/NHQBMCYN.js";
import "../__chunks/GROIW2U2.js";
import "../__chunks/C7ZLNJMM.js";
import "../__chunks/SWN3JYXT.js";
import {
  useStoreState
} from "../__chunks/XYGGOF6W.js";
import "../__chunks/4ODJH2Y5.js";
import {
  createElement,
  createHook,
  forwardRef
} from "../__chunks/ZNLX6YSC.js";
import "../__chunks/G47ZC7SD.js";
import "../__chunks/YXGXYGQX.js";

// src/form/form-submit.tsx
import { invariant } from "@ariakit/core/utils/misc";
var TagName = "button";
var useFormSubmit = createHook(
  function useFormSubmit2({ store, accessibleWhenDisabled = true, ...props }) {
    const context = useFormContext();
    store = store || context;
    invariant(
      store,
      process.env.NODE_ENV !== "production" && "FormSubmit must be wrapped in a Form component."
    );
    props = {
      type: "submit",
      disabled: useStoreState(store, "submitting"),
      ...props
    };
    props = useButton({ ...props, accessibleWhenDisabled });
    return props;
  }
);
var FormSubmit = forwardRef(function FormSubmit2(props) {
  const htmlProps = useFormSubmit(props);
  return createElement(TagName, htmlProps);
});
export {
  FormSubmit,
  useFormSubmit
};
