// @ariakit/react-core/esm/form/form-reset.js
"use client";
import {
  useFormContext
} from "../__chunks/5QCILLRJ.js";
import {
  useButton
} from "../__chunks/NHQBMCYN.js";
import "../__chunks/GROIW2U2.js";
import "../__chunks/C7ZLNJMM.js";
import "../__chunks/SWN3JYXT.js";
import {
  useStoreState
} from "../__chunks/XYGGOF6W.js";
import "../__chunks/4ODJH2Y5.js";
import {
  createElement,
  createHook,
  forwardRef
} from "../__chunks/ZNLX6YSC.js";
import "../__chunks/G47ZC7SD.js";
import "../__chunks/YXGXYGQX.js";

// src/form/form-reset.tsx
import { invariant } from "@ariakit/core/utils/misc";
var TagName = "button";
var useFormReset = createHook(
  function useFormReset2({ store, ...props }) {
    const context = useFormContext();
    store = store || context;
    invariant(
      store,
      process.env.NODE_ENV !== "production" && "FormReset must be wrapped in a Form component."
    );
    props = {
      type: "reset",
      disabled: useStoreState(store, "submitting"),
      ...props
    };
    props = useButton(props);
    return props;
  }
);
var FormReset = forwardRef(function FormReset2(props) {
  const htmlProps = useFormReset(props);
  return createElement(TagName, htmlProps);
});
export {
  FormReset,
  useFormReset
};
