// @wordpress/components/build/private-apis.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/components/src/private-apis.ts
var private_apis_exports = {};
__export(private_apis_exports, {
  privateApis: () => privateApis
});
module.exports = __toCommonJS(private_apis_exports);
var import_react = require("@use-gesture/react");
var import_utils = require("./popover/utils.cjs");
var import_menu = require("./menu/index.cjs");
var import_context_system_provider = require("./context/context-system-provider.cjs");
var import_theme = __toESM(require("./theme/index.cjs"));
var import_tabs = require("./tabs/index.cjs");
var import_strings = require("./utils/strings.cjs");
var import_with_ignore_ime_events = require("./utils/with-ignore-ime-events.cjs");
var import_lock_unlock = require("./lock-unlock.cjs");
var import_badge = __toESM(require("./badge/index.cjs"));
var import_calendar = require("./calendar/index.cjs");
var import_validated_form_controls = require("./validated-form-controls/index.cjs");
var import_form_token_field = require("./validated-form-controls/components/form-token-field.cjs");
var privateApis = {};
(0, import_lock_unlock.lock)(privateApis, {
  __experimentalPopoverLegacyPositionToPlacement: import_utils.positionToPlacement,
  ComponentsContext: import_context_system_provider.ComponentsContext,
  Tabs: import_tabs.Tabs,
  Theme: import_theme.default,
  Menu: import_menu.Menu,
  kebabCase: import_strings.kebabCase,
  withIgnoreIMEEvents: import_with_ignore_ime_events.withIgnoreIMEEvents,
  Badge: import_badge.default,
  normalizeTextString: import_strings.normalizeTextString,
  DateCalendar: import_calendar.DateCalendar,
  DateRangeCalendar: import_calendar.DateRangeCalendar,
  TZDate: import_calendar.TZDate,
  useDrag: import_react.useDrag,
  ValidatedInputControl: import_validated_form_controls.ValidatedInputControl,
  ValidatedCheckboxControl: import_validated_form_controls.ValidatedCheckboxControl,
  ValidatedComboboxControl: import_validated_form_controls.ValidatedComboboxControl,
  ValidatedNumberControl: import_validated_form_controls.ValidatedNumberControl,
  ValidatedSelectControl: import_validated_form_controls.ValidatedSelectControl,
  ValidatedRadioControl: import_validated_form_controls.ValidatedRadioControl,
  ValidatedTextControl: import_validated_form_controls.ValidatedTextControl,
  ValidatedTextareaControl: import_validated_form_controls.ValidatedTextareaControl,
  ValidatedToggleControl: import_validated_form_controls.ValidatedToggleControl,
  ValidatedToggleGroupControl: import_validated_form_controls.ValidatedToggleGroupControl,
  ValidatedFormTokenField: import_form_token_field.ValidatedFormTokenField
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  privateApis
});
