// @wordpress/components/build/menu/checkbox-item.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/components/src/menu/checkbox-item.tsx
var checkbox_item_exports = {};
__export(checkbox_item_exports, {
  CheckboxItem: () => CheckboxItem2
});
module.exports = __toCommonJS(checkbox_item_exports);
var Ariakit = __toESM(require("@ariakit/react"));
var import_element = require("@wordpress/element");
var import_icons = require("@wordpress/icons");
var import_context = require("./context.cjs");
var Styled = __toESM(require("./styles.cjs"));
var import_jsx_runtime = require("react/jsx-runtime");
var CheckboxItem2 = (0, import_element.forwardRef)(function CheckboxItem3({
  suffix,
  children,
  disabled = false,
  hideOnClick = false,
  ...props
}, ref) {
  const menuContext = (0, import_element.useContext)(import_context.Context);
  if (!menuContext?.store) {
    throw new Error("Menu.CheckboxItem can only be rendered inside a Menu component");
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(Styled.CheckboxItem, {
    ref,
    ...props,
    accessibleWhenDisabled: true,
    disabled,
    hideOnClick,
    store: menuContext.store,
    children: [/* @__PURE__ */ (0, import_jsx_runtime.jsx)(Ariakit.MenuItemCheck, {
      store: menuContext.store,
      render: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(Styled.ItemPrefixWrapper, {}),
      style: {
        width: "auto",
        height: "auto"
      },
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_icons.Icon, {
        icon: import_icons.check,
        size: 24
      })
    }), /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(Styled.ItemContentWrapper, {
      children: [/* @__PURE__ */ (0, import_jsx_runtime.jsx)(Styled.ItemChildrenWrapper, {
        children
      }), suffix && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(Styled.ItemSuffixWrapper, {
        children: suffix
      })]
    })]
  });
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  CheckboxItem
});
