// @wordpress/components/build-module/color-palette/utils.mjs
// packages/components/src/color-palette/utils.ts
import { colord, extend } from "colord";
import namesPlugin from "colord/plugins/names";
import a11yPlugin from "colord/plugins/a11y";
import { __ } from "@wordpress/i18n";
extend([namesPlugin, a11yPlugin]);
var isSimpleCSSColor = (value) => {
  const valueIsCssVariable = /var\(/.test(value ?? "");
  const valueIsColorMix = /color-mix\(/.test(value ?? "");
  return !valueIsCssVariable && !valueIsColorMix;
};
var extractColorNameFromCurrentValue = (currentValue, colors = [], showMultiplePalettes = false) => {
  if (!currentValue) {
    return "";
  }
  const currentValueIsSimpleColor = currentValue ? isSimpleCSSColor(currentValue) : false;
  const normalizedCurrentValue = currentValueIsSimpleColor ? colord(currentValue).toHex() : currentValue;
  const colorPalettes = showMultiplePalettes ? colors : [{
    colors
  }];
  for (const {
    colors: paletteColors
  } of colorPalettes) {
    for (const {
      name: colorName,
      color: colorValue
    } of paletteColors) {
      const normalizedColorValue = currentValueIsSimpleColor ? colord(colorValue).toHex() : colorValue;
      if (normalizedCurrentValue === normalizedColorValue) {
        return colorName;
      }
    }
  }
  return __("Custom");
};
var isMultiplePaletteObject = (obj) => Array.isArray(obj.colors) && !("color" in obj);
var isMultiplePaletteArray = (arr) => {
  return arr.length > 0 && arr.every((colorObj) => isMultiplePaletteObject(colorObj));
};
var normalizeColorValue = (value, element) => {
  if (!value || !element || isSimpleCSSColor(value)) {
    return value;
  }
  const {
    ownerDocument
  } = element;
  const {
    defaultView
  } = ownerDocument;
  const computedBackgroundColor = defaultView?.getComputedStyle(element).backgroundColor;
  return computedBackgroundColor ? colord(computedBackgroundColor).toHex() : value;
};
export {
  extractColorNameFromCurrentValue,
  isMultiplePaletteArray,
  isMultiplePaletteObject,
  normalizeColorValue
};
