// @ariakit/react-core/cjs/select/select-label.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _QP4OCRUYcjs = require('../__chunks/QP4OCRUY.cjs');
require('../__chunks/D4JBVI4F.cjs');
require('../__chunks/EZM2TZNU.cjs');
require('../__chunks/PWFJUQMF.cjs');
require('../__chunks/PZ3DLQZR.cjs');
require('../__chunks/YUBPT7BY.cjs');





var _QYN73JUBcjs = require('../__chunks/QYN73JUB.cjs');




var _HAHC3O6Rcjs = require('../__chunks/HAHC3O6R.cjs');
require('../__chunks/XMDZRF6Y.cjs');

// src/select/select-label.tsx
var _misc = require('@ariakit/core/utils/misc');
var TagName = "div";
var useSelectLabel = _QYN73JUBcjs.createHook.call(void 0, 
  function useSelectLabel2({ store, ...props }) {
    const context = _QP4OCRUYcjs.useSelectProviderContext.call(void 0, );
    store = store || context;
    _misc.invariant.call(void 0, 
      store,
      process.env.NODE_ENV !== "production" && "SelectLabel must receive a `store` prop or be wrapped in a SelectProvider component."
    );
    const id = _HAHC3O6Rcjs.useId.call(void 0, props.id);
    const onClickProp = props.onClick;
    const onClick = _HAHC3O6Rcjs.useEvent.call(void 0, (event) => {
      onClickProp == null ? void 0 : onClickProp(event);
      if (event.defaultPrevented) return;
      queueMicrotask(() => {
        const select = store == null ? void 0 : store.getState().selectElement;
        select == null ? void 0 : select.focus();
      });
    });
    props = {
      ...props,
      id,
      ref: _HAHC3O6Rcjs.useMergeRefs.call(void 0, store.setLabelElement, props.ref),
      onClick,
      style: {
        cursor: "default",
        ...props.style
      }
    };
    return _misc.removeUndefinedValues.call(void 0, props);
  }
);
var SelectLabel = _QYN73JUBcjs.memo.call(void 0, 
  _QYN73JUBcjs.forwardRef.call(void 0, function SelectLabel2(props) {
    const htmlProps = useSelectLabel(props);
    return _QYN73JUBcjs.createElement.call(void 0, TagName, htmlProps);
  })
);



exports.SelectLabel = SelectLabel; exports.useSelectLabel = useSelectLabel;
