// @ariakit/react-core/esm/tag/tag-list.js
"use client";
import {
  useTouchDevice
} from "../__chunks/FPWVE2AH.js";
import {
  TagScopedContextProvider,
  useTagProviderContext
} from "../__chunks/3X3QYQCU.js";
import {
  useComposite
} from "../__chunks/TQQOMENK.js";
import "../__chunks/7NJRHOSP.js";
import "../__chunks/57GJCRRF.js";
import "../__chunks/C7ZLNJMM.js";
import "../__chunks/SWN3JYXT.js";
import {
  useStoreState
} from "../__chunks/XYGGOF6W.js";
import "../__chunks/4ODJH2Y5.js";
import {
  createElement,
  createHook,
  forwardRef
} from "../__chunks/ZNLX6YSC.js";
import {
  useEvent,
  useWrapElement
} from "../__chunks/G47ZC7SD.js";
import "../__chunks/YXGXYGQX.js";

// src/tag/tag-list.tsx
import { queueBeforeEvent } from "@ariakit/core/utils/events";
import { getClosestFocusable } from "@ariakit/core/utils/focus";
import { invariant } from "@ariakit/core/utils/misc";
import { isApple } from "@ariakit/core/utils/platform";
import { UndoManager } from "@ariakit/core/utils/undo";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
var TagName = "div";
var useTagList = createHook(
  function useTagList2({ store, ...props }) {
    const context = useTagProviderContext();
    store = store || context;
    invariant(
      store,
      process.env.NODE_ENV !== "production" && "TagList must receive a `store` prop or be wrapped in a TagProvider component."
    );
    const onMouseDownProp = props.onMouseDown;
    const onMouseDown = useEvent((event) => {
      onMouseDownProp == null ? void 0 : onMouseDownProp(event);
      if (event.defaultPrevented) return;
      const target = event.target;
      const currentTarget = event.currentTarget;
      const focusableTarget = getClosestFocusable(target);
      const isSelfFocusable = focusableTarget === currentTarget;
      if (!isSelfFocusable && currentTarget.contains(focusableTarget)) return;
      const { inputElement } = store.getState();
      queueBeforeEvent(event.currentTarget, "mouseup", () => {
        inputElement == null ? void 0 : inputElement.focus();
      });
    });
    const onKeyDownProp = props.onKeyDown;
    const onKeyDown = useEvent((event) => {
      onKeyDownProp == null ? void 0 : onKeyDownProp(event);
      if (event.defaultPrevented) return;
      const pc = !isApple();
      const z = event.key === "z" || event.key === "Z";
      const mod = pc ? event.ctrlKey : event.metaKey;
      const shiftZ = event.shiftKey && z || pc && event.key === "y";
      if (mod && shiftZ) {
        event.preventDefault();
        void UndoManager.redo();
      } else if (mod && z) {
        event.preventDefault();
        void UndoManager.undo();
      }
    });
    props = useWrapElement(
      props,
      (element) => /* @__PURE__ */ jsx(TagScopedContextProvider, { value: store, children: element }),
      [store]
    );
    props = {
      ...props,
      onMouseDown,
      onKeyDown
    };
    props = useComposite({ store, ...props });
    const orientation = useStoreState(
      store,
      (state) => state.orientation === "both" ? void 0 : state.orientation
    );
    const items = useStoreState(store, (state) => state.renderedItems);
    const itemIds = items.filter((item) => !!item.value).map((item) => item.id);
    const labelId = useStoreState(store, (state) => {
      var _a;
      return (_a = state.labelElement) == null ? void 0 : _a.id;
    });
    const listboxProps = {};
    for (const key in props) {
      if (key === "role" || key.startsWith("aria-")) {
        const prop = key;
        listboxProps[prop] = props[prop];
        delete props[prop];
      }
    }
    const touchDevice = useTouchDevice();
    const children = /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsx(
        "div",
        {
          role: touchDevice ? "list" : "listbox",
          "aria-live": "polite",
          "aria-relevant": "all",
          "aria-atomic": true,
          "aria-labelledby": labelId,
          "aria-orientation": orientation,
          "aria-owns": itemIds.join(" "),
          ...listboxProps,
          style: { position: "fixed" }
        }
      ),
      props.children
    ] });
    props = {
      ...props,
      children
    };
    return props;
  }
);
var TagList = forwardRef(function TagList2(props) {
  const htmlProps = useTagList(props);
  return createElement(TagName, htmlProps);
});
export {
  TagList,
  useTagList
};
