// @wordpress/components/build/toolbar/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/components/src/toolbar/index.js
var toolbar_exports = {};
__export(toolbar_exports, {
  Toolbar: () => import_toolbar.default,
  ToolbarButton: () => import_toolbar_button.default,
  ToolbarContext: () => import_toolbar_context.default,
  ToolbarDropdownMenu: () => import_toolbar_dropdown_menu.default,
  ToolbarGroup: () => import_toolbar_group.default,
  ToolbarItem: () => import_toolbar_item.default
});
module.exports = __toCommonJS(toolbar_exports);
var import_toolbar = __toESM(require("./toolbar/index.cjs"));
var import_toolbar_button = __toESM(require("./toolbar-button/index.cjs"));
var import_toolbar_context = __toESM(require("./toolbar-context/index.cjs"));
var import_toolbar_dropdown_menu = __toESM(require("./toolbar-dropdown-menu/index.cjs"));
var import_toolbar_group = __toESM(require("./toolbar-group/index.cjs"));
var import_toolbar_item = __toESM(require("./toolbar-item/index.cjs"));
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Toolbar,
  ToolbarButton,
  ToolbarContext,
  ToolbarDropdownMenu,
  ToolbarGroup,
  ToolbarItem
});
