// @wordpress/rich-text/build/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/rich-text/src/index.ts
var index_exports = {};
__export(index_exports, {
  RichTextData: () => import_create.RichTextData,
  __experimentalRichText: () => __experimentalRichText,
  __unstableCreateElement: () => import_create_element.createElement,
  __unstableToDom: () => import_to_dom.toDom,
  __unstableUseRichText: () => import_hook.__unstableUseRichText,
  applyFormat: () => import_apply_format.applyFormat,
  concat: () => import_concat.concat,
  create: () => import_create.create,
  getActiveFormat: () => import_get_active_format.getActiveFormat,
  getActiveFormats: () => import_get_active_formats.getActiveFormats,
  getActiveObject: () => import_get_active_object.getActiveObject,
  getTextContent: () => import_get_text_content.getTextContent,
  insert: () => import_insert.insert,
  insertObject: () => import_insert_object.insertObject,
  isCollapsed: () => import_is_collapsed.isCollapsed,
  isEmpty: () => import_is_empty.isEmpty,
  join: () => import_join.join,
  privateApis: () => import_private_apis.privateApis,
  registerFormatType: () => import_register_format_type.registerFormatType,
  remove: () => import_remove.remove,
  removeFormat: () => import_remove_format.removeFormat,
  replace: () => import_replace.replace,
  slice: () => import_slice.slice,
  split: () => import_split.split,
  store: () => import_store.store,
  toHTMLString: () => import_to_html_string.toHTMLString,
  toggleFormat: () => import_toggle_format.toggleFormat,
  unregisterFormatType: () => import_unregister_format_type.unregisterFormatType,
  useAnchor: () => import_use_anchor.useAnchor,
  useAnchorRef: () => import_use_anchor_ref.useAnchorRef
});
module.exports = __toCommonJS(index_exports);
var import_store = require("./store/index.cjs");
var import_apply_format = require("./apply-format.cjs");
var import_concat = require("./concat.cjs");
var import_create = require("./create.cjs");
var import_get_active_format = require("./get-active-format.cjs");
var import_get_active_formats = require("./get-active-formats.cjs");
var import_get_active_object = require("./get-active-object.cjs");
var import_get_text_content = require("./get-text-content.cjs");
var import_is_collapsed = require("./is-collapsed.cjs");
var import_is_empty = require("./is-empty.cjs");
var import_join = require("./join.cjs");
var import_register_format_type = require("./register-format-type.cjs");
var import_remove_format = require("./remove-format.cjs");
var import_remove = require("./remove.cjs");
var import_replace = require("./replace.cjs");
var import_insert = require("./insert.cjs");
var import_insert_object = require("./insert-object.cjs");
var import_slice = require("./slice.cjs");
var import_split = require("./split.cjs");
var import_to_dom = require("./to-dom.cjs");
var import_to_html_string = require("./to-html-string.cjs");
var import_toggle_format = require("./toggle-format.cjs");
var import_unregister_format_type = require("./unregister-format-type.cjs");
var import_create_element = require("./create-element.cjs");
var import_private_apis = require("./private-apis.cjs");
var import_use_anchor_ref = require("./hook/use-anchor-ref.cjs");
var import_use_anchor = require("./hook/use-anchor.cjs");
var import_hook = require("./hook/index.cjs");
function __experimentalRichText() {
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  RichTextData,
  __experimentalRichText,
  __unstableCreateElement,
  __unstableToDom,
  __unstableUseRichText,
  applyFormat,
  concat,
  create,
  getActiveFormat,
  getActiveFormats,
  getActiveObject,
  getTextContent,
  insert,
  insertObject,
  isCollapsed,
  isEmpty,
  join,
  privateApis,
  registerFormatType,
  remove,
  removeFormat,
  replace,
  slice,
  split,
  store,
  toHTMLString,
  toggleFormat,
  unregisterFormatType,
  useAnchor,
  useAnchorRef
});
