// @ariakit/react-core/cjs/tooltip/tooltip.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _KADSBCBXcjs = require('../__chunks/KADSBCBX.cjs');


var _MZ6ANGRLcjs = require('../__chunks/MZ6ANGRL.cjs');
require('../__chunks/L6CLLK7R.cjs');
require('../__chunks/4AJ2AADN.cjs');
require('../__chunks/YE42HHT5.cjs');


var _YZCSZZOHcjs = require('../__chunks/YZCSZZOH.cjs');
require('../__chunks/T3NAWF7A.cjs');
require('../__chunks/UI5KWBYV.cjs');
require('../__chunks/L634CRNJ.cjs');
require('../__chunks/XQY3LU3Z.cjs');
require('../__chunks/HST3DJBN.cjs');
require('../__chunks/XINXIOPH.cjs');
require('../__chunks/W32FX7DM.cjs');
require('../__chunks/EQM67Z5W.cjs');
require('../__chunks/6GRO3ZM4.cjs');
require('../__chunks/FSMQ3LQP.cjs');
require('../__chunks/NNIVHDZH.cjs');
require('../__chunks/KPQQ36MY.cjs');
require('../__chunks/JLMLKTUB.cjs');
require('../__chunks/LCWSLOYU.cjs');
require('../__chunks/V7JASX7S.cjs');
require('../__chunks/56NPFF7P.cjs');
require('../__chunks/GASMO7Y4.cjs');
require('../__chunks/DLDJYPHW.cjs');
require('../__chunks/26P4PLHI.cjs');
require('../__chunks/YVOQ2BCB.cjs');
require('../__chunks/4SKDRUBR.cjs');
require('../__chunks/AQ2LO7WV.cjs');
require('../__chunks/OXZTKJQE.cjs');
require('../__chunks/OINITWAS.cjs');
require('../__chunks/NFOCAOK7.cjs');
require('../__chunks/N2EWCFNQ.cjs');
require('../__chunks/D4JBVI4F.cjs');
require('../__chunks/EZM2TZNU.cjs');
require('../__chunks/PWFJUQMF.cjs');
require('../__chunks/ZDWFUT2O.cjs');
require('../__chunks/T3QB4FR3.cjs');


var _JRPZ5DZLcjs = require('../__chunks/JRPZ5DZL.cjs');




var _QYN73JUBcjs = require('../__chunks/QYN73JUB.cjs');


var _HAHC3O6Rcjs = require('../__chunks/HAHC3O6R.cjs');
require('../__chunks/XMDZRF6Y.cjs');

// src/tooltip/tooltip.tsx
var _dom = require('@ariakit/core/utils/dom');
var _misc = require('@ariakit/core/utils/misc');
var _jsxruntime = require('react/jsx-runtime');
var TagName = "div";
var useTooltip = _QYN73JUBcjs.createHook.call(void 0, 
  function useTooltip2({
    store,
    portal = true,
    gutter = 8,
    preserveTabOrder = false,
    hideOnHoverOutside = true,
    hideOnInteractOutside = true,
    ...props
  }) {
    const context = _KADSBCBXcjs.useTooltipProviderContext.call(void 0, );
    store = store || context;
    _misc.invariant.call(void 0, 
      store,
      process.env.NODE_ENV !== "production" && "Tooltip must receive a `store` prop or be wrapped in a TooltipProvider component."
    );
    props = _HAHC3O6Rcjs.useWrapElement.call(void 0, 
      props,
      (element) => /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _KADSBCBXcjs.TooltipScopedContextProvider, { value: store, children: element }),
      [store]
    );
    const role = _JRPZ5DZLcjs.useStoreState.call(void 0, 
      store,
      (state) => state.type === "description" ? "tooltip" : "none"
    );
    props = { role, ...props };
    props = _MZ6ANGRLcjs.useHovercard.call(void 0, {
      ...props,
      store,
      portal,
      gutter,
      preserveTabOrder,
      hideOnHoverOutside(event) {
        if (_misc.isFalsyBooleanCallback.call(void 0, hideOnHoverOutside, event)) return false;
        const anchorElement = store == null ? void 0 : store.getState().anchorElement;
        if (!anchorElement) return true;
        if ("focusVisible" in anchorElement.dataset) return false;
        return true;
      },
      hideOnInteractOutside: (event) => {
        if (_misc.isFalsyBooleanCallback.call(void 0, hideOnInteractOutside, event)) return false;
        const anchorElement = store == null ? void 0 : store.getState().anchorElement;
        if (!anchorElement) return true;
        if (_dom.contains.call(void 0, anchorElement, event.target)) return false;
        return true;
      }
    });
    return props;
  }
);
var Tooltip = _YZCSZZOHcjs.createDialogComponent.call(void 0, 
  _QYN73JUBcjs.forwardRef.call(void 0, function Tooltip2(props) {
    const htmlProps = useTooltip(props);
    return _QYN73JUBcjs.createElement.call(void 0, TagName, htmlProps);
  }),
  _KADSBCBXcjs.useTooltipProviderContext
);



exports.Tooltip = Tooltip; exports.useTooltip = useTooltip;
